
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://joon-klaps.github.io/viralgenie/latest/dev/output/">
      
      
        <link rel="prev" href="../parameters/">
      
      
        <link rel="next" href="../workflow/">
      
      
      <link rel="icon" href="../images/ViralGenie-nf-core-theme.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.4">
    
    
      
        <title>Output - Viralgenie</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:300,300i,400,400i,700,700i%7CUbuntu+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Ubuntu";--md-code-font:"Ubuntu Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/katex.min.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-DJ9PDTBEEC"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-DJ9PDTBEEC",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-DJ9PDTBEEC",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>if("undefined"!=typeof __md_analytics){var consent=__md_get("__consent");consent&&consent.analytics&&__md_analytics()}</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Viralgenie" class="md-header__button md-logo" aria-label="Viralgenie" data-md-component="logo">
      
  <img src="../images/ViralGenie-nf-core-theme.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Viralgenie
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Joon-Klaps/viralgenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Joon-Klaps/viralgenie
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../usage/" class="md-tabs__link">
        
  
    
  
  Usage

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../parameters/" class="md-tabs__link">
        
  
    
  
  Parameters

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  Output

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../workflow/" class="md-tabs__link">
          
  
    
  
  Workflows

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../customisation/" class="md-tabs__link">
          
  
    
  
  Customisation

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../CITATIONS/" class="md-tabs__link">
        
  
    
  
  Citations

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../CONTRIBUTING/" class="md-tabs__link">
        
  
    
  
  Contributing

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
                
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" hidden>
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Viralgenie" class="md-nav__button md-logo" aria-label="Viralgenie" data-md-component="logo">
      
  <img src="../images/ViralGenie-nf-core-theme.png" alt="logo">

    </a>
    Viralgenie
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Joon-Klaps/viralgenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Joon-Klaps/viralgenie
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fastp" class="md-nav__link">
    <span class="md-ellipsis">
      fastp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#trimmomatic" class="md-nav__link">
    <span class="md-ellipsis">
      Trimmomatic
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umi-deduplication" class="md-nav__link">
    <span class="md-ellipsis">
      UMI-deduplication
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bbduk" class="md-nav__link">
    <span class="md-ellipsis">
      BBDuk
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prinseq" class="md-nav__link">
    <span class="md-ellipsis">
      prinseq++
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hostremoval-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Hostremoval-Kraken2
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomic-diversity" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomic Diversity
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Metagenomic Diversity">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Kraken2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju" class="md-nav__link">
    <span class="md-ellipsis">
      Kaiju
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#krona" class="md-nav__link">
    <span class="md-ellipsis">
      Krona
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-polishing" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly &amp; Polishing
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Assembly & Polishing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#assemblers" class="md-nav__link">
    <span class="md-ellipsis">
      Assemblers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sspace-basic" class="md-nav__link">
    <span class="md-ellipsis">
      SSPACE Basic
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prinseq-contigs" class="md-nav__link">
    <span class="md-ellipsis">
      prinseq++ - contigs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blast" class="md-nav__link">
    <span class="md-ellipsis">
      BLAST
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preclustering-kaiju-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Preclustering - Kaiju &amp; Kraken2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering" class="md-nav__link">
    <span class="md-ellipsis">
      Clustering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clustering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cd-hit-est" class="md-nav__link">
    <span class="md-ellipsis">
      CD-HIT-EST
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vsearch-cluster" class="md-nav__link">
    <span class="md-ellipsis">
      vsearch-cluster
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mmseqs2" class="md-nav__link">
    <span class="md-ellipsis">
      MMseqs2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vrhyme" class="md-nav__link">
    <span class="md-ellipsis">
      vRhyme
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mash" class="md-nav__link">
    <span class="md-ellipsis">
      Mash
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#minimap2" class="md-nav__link">
    <span class="md-ellipsis">
      Minimap2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ivar-contig-consensus" class="md-nav__link">
    <span class="md-ellipsis">
      iVar contig consensus
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-calling-iterative-refinement" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling &amp; Iterative Refinement
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variant Calling & Iterative Refinement">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#reference-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Reference selection
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Read mapping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deduplication" class="md-nav__link">
    <span class="md-ellipsis">
      Deduplication
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Deduplication">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#umi-tools" class="md-nav__link">
    <span class="md-ellipsis">
      UMI-tools
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#picard-mark-duplicates" class="md-nav__link">
    <span class="md-ellipsis">
      Picard - Mark Duplicates
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mapping-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Mapping statistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Mapping statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#samtools" class="md-nav__link">
    <span class="md-ellipsis">
      Samtools
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#picard-collect-multiple-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      Picard - Collect Multiple Metrics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-mpileup-like-file" class="md-nav__link">
    <span class="md-ellipsis">
      Custom - mpileup like file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth-coverage" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth - Coverage
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Variant calling &amp; filtering
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#consensus-generation" class="md-nav__link">
    <span class="md-ellipsis">
      Consensus generation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#consensus-quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      Consensus Quality control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Consensus Quality control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#quast" class="md-nav__link">
    <span class="md-ellipsis">
      Quast
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#checkv" class="md-nav__link">
    <span class="md-ellipsis">
      CheckV
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blastn" class="md-nav__link">
    <span class="md-ellipsis">
      BLASTn
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mmseqs-search-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      MMseqs-search (annotation)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mafft" class="md-nav__link">
    <span class="md-ellipsis">
      MAFFT
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline information
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../workflow/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Workflows
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_7" id="__nav_7_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Workflows
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/preprocessing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Preprocessing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/metagenomic_diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metagenomic diversity
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/assembly_polishing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Assembly & Polishing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/variant_and_refinement/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant analysis & Iterative refinement
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/consensus_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Consensus evaluation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../customisation/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Customisation
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_8" id="__nav_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Customisation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../customisation/databases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Databases
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../customisation/configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../CITATIONS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Citations
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../CONTRIBUTING/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
                
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#fastqc" class="md-nav__link">
    <span class="md-ellipsis">
      FastQC
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#fastp" class="md-nav__link">
    <span class="md-ellipsis">
      fastp
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#trimmomatic" class="md-nav__link">
    <span class="md-ellipsis">
      Trimmomatic
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#umi-deduplication" class="md-nav__link">
    <span class="md-ellipsis">
      UMI-deduplication
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bbduk" class="md-nav__link">
    <span class="md-ellipsis">
      BBDuk
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prinseq" class="md-nav__link">
    <span class="md-ellipsis">
      prinseq++
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#hostremoval-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Hostremoval-Kraken2
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metagenomic-diversity" class="md-nav__link">
    <span class="md-ellipsis">
      Metagenomic Diversity
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Metagenomic Diversity">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Kraken2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kaiju" class="md-nav__link">
    <span class="md-ellipsis">
      Kaiju
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#krona" class="md-nav__link">
    <span class="md-ellipsis">
      Krona
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assembly-polishing" class="md-nav__link">
    <span class="md-ellipsis">
      Assembly &amp; Polishing
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Assembly & Polishing">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#assemblers" class="md-nav__link">
    <span class="md-ellipsis">
      Assemblers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sspace-basic" class="md-nav__link">
    <span class="md-ellipsis">
      SSPACE Basic
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prinseq-contigs" class="md-nav__link">
    <span class="md-ellipsis">
      prinseq++ - contigs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blast" class="md-nav__link">
    <span class="md-ellipsis">
      BLAST
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#preclustering-kaiju-kraken2" class="md-nav__link">
    <span class="md-ellipsis">
      Preclustering - Kaiju &amp; Kraken2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering" class="md-nav__link">
    <span class="md-ellipsis">
      Clustering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clustering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cd-hit-est" class="md-nav__link">
    <span class="md-ellipsis">
      CD-HIT-EST
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vsearch-cluster" class="md-nav__link">
    <span class="md-ellipsis">
      vsearch-cluster
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mmseqs2" class="md-nav__link">
    <span class="md-ellipsis">
      MMseqs2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#vrhyme" class="md-nav__link">
    <span class="md-ellipsis">
      vRhyme
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mash" class="md-nav__link">
    <span class="md-ellipsis">
      Mash
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#minimap2" class="md-nav__link">
    <span class="md-ellipsis">
      Minimap2
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ivar-contig-consensus" class="md-nav__link">
    <span class="md-ellipsis">
      iVar contig consensus
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-calling-iterative-refinement" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Calling &amp; Iterative Refinement
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Variant Calling & Iterative Refinement">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#reference-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Reference selection
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      Read mapping
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deduplication" class="md-nav__link">
    <span class="md-ellipsis">
      Deduplication
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Deduplication">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#umi-tools" class="md-nav__link">
    <span class="md-ellipsis">
      UMI-tools
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#picard-mark-duplicates" class="md-nav__link">
    <span class="md-ellipsis">
      Picard - Mark Duplicates
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mapping-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Mapping statistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Mapping statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#samtools" class="md-nav__link">
    <span class="md-ellipsis">
      Samtools
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#picard-collect-multiple-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      Picard - Collect Multiple Metrics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#custom-mpileup-like-file" class="md-nav__link">
    <span class="md-ellipsis">
      Custom - mpileup like file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mosdepth-coverage" class="md-nav__link">
    <span class="md-ellipsis">
      Mosdepth - Coverage
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-calling-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      Variant calling &amp; filtering
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#consensus-generation" class="md-nav__link">
    <span class="md-ellipsis">
      Consensus generation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#consensus-quality-control" class="md-nav__link">
    <span class="md-ellipsis">
      Consensus Quality control
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Consensus Quality control">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#quast" class="md-nav__link">
    <span class="md-ellipsis">
      Quast
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#checkv" class="md-nav__link">
    <span class="md-ellipsis">
      CheckV
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blastn" class="md-nav__link">
    <span class="md-ellipsis">
      BLASTn
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mmseqs-search-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      MMseqs-search (annotation)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mafft" class="md-nav__link">
    <span class="md-ellipsis">
      MAFFT
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multiqc" class="md-nav__link">
    <span class="md-ellipsis">
      MultiQC
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-information" class="md-nav__link">
    <span class="md-ellipsis">
      Pipeline information
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  
  



  
  


<h1 id="output">Output</h1>
<h2 id="introduction">Introduction</h2>
<p>This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarizes results at the end of the pipeline.</p>
<p>The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>A global, partly random prefix can be created using the argument <code>--prefix &lt;string&gt;</code>. The following string will then be used as a prefix to all output files:
<div class="language-java highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="s">&quot;&lt;prefix_string&gt;_&lt;date&gt;_&lt;pipeline_version&gt;_&lt;workflow_runName&gt;&quot;</span>
</span></code></pre></div></p>
</div>
<h2 id="preprocessing">Preprocessing</h2>
<p>All output files of the preprocessing steps can be found in the directory <code>preprocessing/</code>.</p>
<h3 id="fastqc">FastQC</h3>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>fastqc/{raw,trim,host}</code><ul>
<li><code>*_fastqc.html</code>: FastQC report containing quality metrics.</li>
<li><code>*_fastqc.zip</code>: Zip archive containing the FastQC report, tab-delimited data file, and plot images.</li>
</ul>
</li>
</ul>
</details>
<p><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination, and overrepresented sequences. For further reading and documentation see the <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/">FastQC help pages</a>.</p>
<p><img alt="MultiQC - FastQC sequence counts plot" src="../images/mqc_fastqc_counts.png" /></p>
<p><img alt="MultiQC - FastQC mean quality scores plot" src="../images/mqc_fastqc_quality.png" /></p>
<p><img alt="MultiQC - FastQC adapter content plot" src="../images/mqc_fastqc_adapter.png" /></p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The FastQC plots displayed in the MultiQC report show <em>untrimmed</em>, <em>trimmed</em>, and <em>host filtered</em> reads. Make sure to check the section titles for the correct set of reads.</p>
</div>
<h3 id="fastp">fastp</h3>
<p><a href="https://github.com/OpenGene/fastp">fastp</a> is a FASTQ pre-processing tool for quality control, trimming of adapters, quality filtering, and other features.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>fastp/</code><ul>
<li><code>report/&lt;sample-id&gt;.*{html,json}</code>: report files in different formats.</li>
<li><code>log/&lt;sample-id&gt;.*{html,json}</code>: log files.</li>
<li><code>&lt;sample-id&gt;.fastp.fastq.gz</code>: file with the trimmed fastq reads.</li>
<li><code>fail/&lt;sample-id&gt;.fail.fastq.gz</code>: file with reads that didn't suffice quality controls.</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will only provide the report and log files if fastp is selected. The trimmed reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'trimming'</code>. Similarly, the saving of the output reads can be enabled with <code>--save_trimmed_fail</code>.</p>
<h3 id="trimmomatic">Trimmomatic</h3>
<p><a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a> is a FASTQ pre-processing tool for quality control, trimming of adapters, quality filtering, and other features.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>trimmomatic/</code><ul>
<li><code>&lt;sample-id&gt;.fastq.gz</code>: file with the trimmed fastq reads.</li>
<li><code>log/&lt;sample-id&gt;.*{html,txt,zip}</code>: log files generated by Trimmomatic.</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will only provide the report and log files if Trimmomatic is selected. The trimmed reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'trimming'</code>.</p>
<h3 id="umi-deduplication">UMI-deduplication</h3>
<p>UMI-deduplication can be done at the read level using <a href="https://humid.readthedocs.io/en/latest/usage.html"><code>HUMID</code></a>. Viralgenie also provides the opportunity to extract the UMI from the read using <a href="https://umi-tools.readthedocs.io/en/latest/QUICK_START.html#step-3--extract-the-umis"><code>UMI-tools extract</code></a> if the UMI is not in the header. Results will be stored in the <code>preprocessing/umi</code> directory.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>umi/</code><ul>
<li><code>humid/</code><ul>
<li><code>log/&lt;sample-id&gt;.log</code>: log file of HUMID.</li>
<li><code>annotated/&lt;sample-id&gt;_annotated_*.fastq.gz</code>: annotated FastQ files, reads will have their assigned cluster in the read header.</li>
<li><code>deduplicated/&lt;sample-id&gt;_deduplicated_*.fastq.gz</code>: deduplicated FastQ files.</li>
</ul>
</li>
<li><code>umitools/</code><ul>
<li><code>log/&lt;sample-id&gt;.log</code>: log file of UMI-tools.</li>
<li><code>extracts/&lt;sample-id&gt;.umi_extract*.fastq.gz</code>: fastq file where UMIs have been removed from the read and moved to the read header.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will not assume reads have UMIs. To enable this, use the parameter <code>--with_umi</code>. Specify where UMI deduplication should occur with <code>--umi_deduplicate</code> if at a <code>read</code> level, on a <code>mapping</code> level, or <code>both</code> at a read and mapping level. The deduplicated reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'deduplication'</code>.</p>
<h3 id="bbduk">BBDuk</h3>
<p><a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/">BBDuk</a> stands for Decontamination Using Kmers. BBDuk was developed to combine most common data-quality-related trimming, filtering, and masking operations into a single high-performance tool.</p>
<p>It is used in viralgenie for complexity filtering using different algorithms. This means that it will remove reads with low sequence diversity (e.g., mono- or dinucleotide repeats).</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>bbduk/</code><ul>
<li><code>log/&lt;sample-id&gt;.bbduk.log</code>: log file containing filtering statistics.</li>
<li><code>&lt;sample-id&gt;.fastq.gz</code>: resulting FASTQ file without low-complexity reads.</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will only provide the log files of BBDuk. The filtered reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'complexity'</code>.</p>
<h3 id="prinseq">prinseq++</h3>
<p><a href="https://github.com/Adrian-Cantu/PRINSEQ-plus-plus"><code>prinseq++</code></a> is used in viralgenie for complexity filtering using different algorithms. This means that it will remove reads with low sequence diversity (e.g., mono- or dinucleotide repeats).</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>prinseq/</code><ul>
<li><code>log/&lt;sample-id&gt;.log</code>: log file containing filtering statistics.</li>
<li><code>&lt;sample-id&gt;.fastq.gz</code>: resulting FASTQ file without low-complexity reads.</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will only provide the log files of prinseq. The filtered reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'complexity'</code>.</p>
<h3 id="hostremoval-kraken2">Hostremoval-Kraken2</h3>
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps k-mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>hostremoval-kraken2/</code><ul>
<li><code>&lt;sample-id&gt;_kraken2_host.report.txt</code>: A profile of the aligned reads to a given host contamination database.</li>
<li><code>&lt;sample-id&gt;_kraken2_host.unclassified*.fastq.gz</code>: resulting FASTQ file with reads that don't have any matches to the given host contamination database.</li>
</ul>
</li>
</ul>
</details>
<p>By default, viralgenie will only provide the log files of Kraken2 which are visualized in <a href="#multiqc">MultiQC</a>. The filtered reads can be saved by specifying <code>--save_intermediate_reads</code> or <code>--save_final_reads 'host'</code>.</p>
<h2 id="metagenomic-diversity">Metagenomic Diversity</h2>
<p>The results of the metagenomic diversity analysis are stored in the directory <code>metagenomic_diversity/</code>. Results are also visualized in the MultiQC report.
<img alt="MultiQC kaiju example" src="../images/mqc_kaiju.png" /></p>
<h3 id="kraken2">Kraken2</h3>
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps k-mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>metagenomic_diversity/kraken2/</code><ul>
<li><code>&lt;sample-id&gt;.report.txt</code>: A Kraken2 report that summarizes the fraction abundance, taxonomic ID, number of k-mers, taxonomic path of all the hits in the Kraken2 run for a given sample. Will be 6 columns rather than 8 if <code>--save_minimizers</code> specified.</li>
<li><code>&lt;sample-id&gt;_kraken2_host.unclassified*.fastq.gz</code>: resulting FASTQ file with reads that don't have any matches to the given host contamination database.</li>
<li><code>&lt;sample-id&gt;.classified.fastq.gz</code>: FASTQ file containing all reads that had a hit against a reference in the database for a given sample.</li>
<li><code>&lt;sample-id&gt;.unclassified.fastq.gz</code>: FASTQ file containing all reads that did not have a hit in the database for a given sample.</li>
<li><code>&lt;sample-id&gt;.classifiedreads.txt</code>: A list of read IDs and the hits each read had against the database for a given sample.</li>
</ul>
</li>
</ul>
</details>
<blockquote>
<p>By default, viralgenie will provide any classified or unclassified fastq files, specify this with <code>--kraken2_save_reads</code>. Similarly, for the classified reads table, specify this with <code>--kraken2_save_readclassification</code>.</p>
</blockquote>
<h3 id="kaiju">Kaiju</h3>
<p><a href="https://kaiju.binf.ku.dk/">Kaiju</a> is a program for sensitive taxonomic classification of high-throughput sequencing reads from metagenomic data. It is based on the Burrows-Wheeler transform and the lowest common ancestor algorithm.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>metagenomic_diversity/kaiju/</code><ul>
<li><code>&lt;sample-id&gt;.tsv</code>: Raw output from Kaiju with taxonomic rank, read ID, and taxonomic ID.</li>
<li><code>&lt;sample-id&gt;.txt</code>: A summary of the taxonomic classification of the reads in the sample.</li>
</ul>
</li>
</ul>
</details>
<h3 id="krona">Krona</h3>
<p><a href="https://github.com/marbl/Krona/wiki">Krona</a> is a hierarchical data visualization tool that can be used to visualize the taxonomic classification of metagenomic data.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>metagenomic_diversity/krona/</code><ul>
<li><code>&lt;kaiju|kraken2&gt;_.html</code>: A HTML file containing the Krona visualization of the taxonomic classification of the reads in the sample.</li>
</ul>
</li>
</ul>
</details>
<p><img alt="Krona example COVID" class="center" src="../images/krona_covid.png" /></p>
<h2 id="assembly-polishing">Assembly &amp; Polishing</h2>
<p>The results of the assembly processes &amp; polishing are stored in the directory <code>assembly/</code>.</p>
<p>Multiple intermediate files can be generated during the assembly process, some of them might not always be interesting to have. For this reason, there is an option to save the intermediate files with the <code>--save_intermediate_polishing</code> argument which is by default off.</p>
<h3 id="assemblers">Assemblers</h3>
<p>Multiple assemblers [spades, trinity, megahit] can be used which have their results combined. Each assembler has its own directory in the <code>assembly/assemblers</code> directory, where there will be a subfolder for the contigs and the QC results from QUAST.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>assemblers/</code><ul>
<li><code>spades/&lt;spades_mode&gt;/</code><ul>
<li><code>contigs/&lt;sample-id&gt;_spades.fa.gz</code>: Contigs generated by SPAdes.</li>
<li><code>log/&lt;sample-id&gt;_spades.log</code>: Directory containing the log file of the SPAdes run.</li>
<li><code>quast/&lt;sample-id&gt;_spades.tsv</code>: Directory containing the QUAST report.</li>
</ul>
</li>
<li><code>trinity/</code><ul>
<li><code>contigs/&lt;sample-id&gt;_trinity.fa.gz</code>: Contigs generated by Trinity.</li>
<li><code>quast/&lt;sample-id&gt;_trinity.tsv</code>: Directory containing the QUAST report.</li>
</ul>
</li>
<li><code>megahit/</code><ul>
<li><code>contigs/&lt;sample-id&gt;_megahit.fa.gz</code>: Contigs generated by Megahit.</li>
<li><code>quast/&lt;sample-id&gt;_megahit.tsv</code>: Directory containing the QUAST report.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<p>QUAST results are also summarized and plotted in the MultiQC report.</p>
<p><img alt="MultiQC - QUAST assembly statistics" class="center" src="../images/mqc_quast.png" /></p>
<p>Finally, the results of the assemblers are combined and stored in the <code>tools_combined/</code> directory.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>assemblers</code><ul>
<li><code>tools_combined/&lt;sample-id&gt;.combined.fa</code>: Contigs generated by combining the results of the assemblers.</li>
</ul>
</li>
</ul>
</details>
<h3 id="sspace-basic">SSPACE Basic</h3>
<p><a href="https://github.com/nsoranzo/sspace_basic">SSPACE Basic</a> is a tool for scaffolding contigs using paired-end reads. It is modified from the SSAKE assembler and has the feature of extending contigs using reads that are unmappable in the contig assembly step.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>sspace_basic/</code><ul>
<li><code>scaffolds/&lt;sample-id&gt;.scaffolds.fasta</code>: Scaffolds generated by SSPACE Basic.</li>
<li><code>log/&lt;sample-id&gt;.*.txt</code>: Various txt files containing log and summary information on the SSPACE Basic run.</li>
</ul>
</li>
</ul>
</details>
<h3 id="prinseq-contigs">prinseq++ - contigs</h3>
<p><a href="https://github.com/Adrian-Cantu/PRINSEQ-plus-plus"><code>prinseq++</code></a> is used for complexity filtering of contigs.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>prinseq/</code><ul>
<li><code>scaffolds/&lt;sample-id&gt;.scaffolds.fasta</code>: Scaffolds generated by SSPACE Basic.</li>
<li><code>log/&lt;sample-id&gt;.*.txt</code>: Various txt files containing log and summary information on the SSPACE Basic run.</li>
</ul>
</li>
</ul>
</details>
<h3 id="blast">BLAST</h3>
<p><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> is a sequence comparison tool that can be used to compare a query sequence against a database of sequences. In viralgenie, BLAST is used to compare the contigs generated by the assemblers to a database of viral sequences.</p>
<p>By default, viralgenie will only provide the BLAST results in a tabular format. It will have selected only for the top five hits and will also have a filtered version where it will only include hits with an e-value of 0.01 or lower, a bitscore of 50 or higher, and an alignment percentage of 0.80 or higher.</p>
<details class="- info">
<summary>Column names</summary>
<ul>
<li>qseqid</li>
<li>sseqid</li>
<li>stitle</li>
<li>pident</li>
<li>qlen</li>
<li>slen</li>
<li>length</li>
<li>mismatch</li>
<li>gapopen</li>
<li>qstart</li>
<li>qend</li>
<li>sstart</li>
<li>send</li>
<li>evalue</li>
<li>bitscore</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/</code><ul>
<li><code>blast/&lt;sample-id&gt;_filter.tsv</code>: Filtered BLAST results in tabular format.</li>
<li><code>intermediate/blast/filtered-sequences/&lt;sample-id&gt;_withref.fa</code>: Contigs with the blast hit sequence in a fasta file.</li>
<li><code>intermediate/blast/hits/&lt;sample-id&gt;.txt</code>: unfiltered BLAST results in tabular format.</li>
</ul>
</li>
</ul>
</details>
<blockquote>
<p>By default, viralgenie will only provide the filtered blast.txt file. The intermediate files can be saved by specifying <code>--save_intermediate_polishing</code>.</p>
</blockquote>
<h3 id="preclustering-kaiju-kraken2">Preclustering - Kaiju &amp; Kraken2</h3>
<p><a href="https://kaiju.binf.ku.dk/">Kaiju</a> is a program for sensitive taxonomic classification of high-throughput sequencing reads from metagenomic data. It is based on the Burrows-Wheeler transform and the lowest common ancestor algorithm.</p>
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps k-mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/intermediate/precluster</code><ul>
<li><code>kaiju/&lt;sample-id&gt;_kaiju.tsv</code>: Raw output from Kaiju with taxonomic rank, read ID, and taxonomic ID.</li>
<li><code>kraken2/&lt;sample-id&gt;_kraken2_reports.txt</code>: A Kraken2 report that summarizes the fraction abundance, taxonomic ID, number of k-mers, taxonomic path of all the hits in the Kraken2 run for a given sample. Will be 6 columns rather than 8 if <code>--save_minimizers</code> specified.</li>
<li><code>kraken2/&lt;sample-id&gt;_kraken2.classifiedreads.txt</code>: A list of read IDs and the hits each read had against the database for a given sample.</li>
<li><code>merged_classifications/&lt;sample-id&gt;.txt</code>: Taxonomy merged based on the specified strategy, filtered based on specified filters, and simplified up to a certain taxonomy with the columns being taxonomic rank, read ID, and taxonomic ID.</li>
<li><code>sequences/&lt;sample-id&gt;/&lt;sample-id&gt;_taxid&lt;taxonomic ID&gt;.fa</code>: Fasta file with the contigs that were classified to that specific taxonomic ID.</li>
</ul>
</li>
</ul>
</details>
<blockquote>
<p>By default, viralgenie will not provide any preclustering files. The intermediate files can be saved by specifying <code>--save_intermediate_polishing</code>.</p>
</blockquote>
<h3 id="clustering">Clustering</h3>
<p>The output files of each clustering method are directly put in the <code>assembly/polishing</code> directory, with the exception of a summary file that is generated by the pipeline for each cluster with the size of the cluster, the centroid, etc.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/intermediate/cluster/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;.summary_mqc.tsv</code>: A tabular file with comments used for <a href="#multiqc">MultiQC</a> with statistics on the number of identified clusters in a sample.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;.clusters.tsv</code>: A tabular file with metadata on all clusters in a sample. It's the JSON file of all clusters in a table format.</li>
</ul>
</li>
</ul>
</details>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Whenever there is a 'cl#' in the file name, it refers to the cluster number of that sample.</p>
</div>
<blockquote>
<p>By default, viralgenie will not provide any clustering overview files. The intermediate files can be saved by specifying <code>--save_intermediate_polishing</code>.</p>
</blockquote>
<h4 id="cd-hit-est">CD-HIT-EST</h4>
<p><a href="https://github.com/weizhongli/cdhit">CD-HIT</a> is a very fast, widely used program for clustering and comparing protein or nucleotide sequences.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/cdhit/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;.fa.clstr</code>: A cluster file containing the clustering information. where  "&gt;" starts a new cluster, a "*" at the end means that this sequence is the representative or centroid of this cluster, and a "%" is the identity between this sequence and the representative.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;.fa</code>: A fasta file containing the centroid sequence.</li>
</ul>
</li>
</ul>
</details>
<h4 id="vsearch-cluster">vsearch-cluster</h4>
<p><a href="https://github.com/torognes/vsearch">vsearch</a> implements a single-pass, greedy centroid-based clustering algorithm, similar to the algorithms implemented in usearch, DNAclust, and sumaclust for example. The output has to be in the <code>--uc</code> format or else the pipeline will not be able to process the output.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/vsearch/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;.tsv.gz</code>: A cluster file containing the clustering information.</li>
</ul>
</li>
</ul>
</details>
<details class="- info">
<summary>vsearch -uc columns</summary>
<ol>
<li>Entry (S, H, or C):</li>
<li>Record type: S, H, or C.</li>
<li>Cluster number (zero-based).</li>
<li>Centroid length (S), query length (H), or cluster size (C).</li>
<li>Percentage of similarity with the centroid sequence (H), or set to * (S, C).</li>
<li>Match orientation + or - (H), or set to <em> (S, C). Not used, always set to </em> (S, C) or to zero (H).</li>
<li>Not used, always set to * (S, C) or to zero (H).</li>
<li>Set to * (S, C) or, for H, compact representation of the pairwise alignment using the CIGAR format (Compact Idiosyncratic Gapped Alignment Report): M (match/mismatch), D (deletion), and I (insertion). The equal sign = indicates that the query is identical to the centroid sequence.</li>
<li>Label of the query sequence (H), or of the centroid sequence (S, C). 10. Label of the centroid sequence (H), or set to * (S, C).</li>
</ol>
</details>
<h4 id="mmseqs2">MMseqs2</h4>
<p><a href="https://github.com/soedinglab/MMseqs2/wiki">MMseqs2</a> is a software suite to search and cluster huge protein and nucleotide sequence sets. The cascaded clustering workflow (<code>mmseqs-cluster</code>) first runs linclust, the linear-time clustering module of mmseqs (<code>mmseqs-linclust</code>), that can produce clusterings down to 50% sequence identity in very short time.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/</code><ul>
<li><code>mmseqs2/&lt;sample-id&gt;/&lt;sample-id&gt;.tsv</code>: A cluster file containing the clustering information. Where the first column is the cluster representative and the second column the member.</li>
<li><code>intermediate/mmseqs/clustered_db/&lt;sample-id&gt;*</code>: A MMseqs2 database of the clustered sequences.</li>
<li><code>intermediate/mmseqs/sequence_db/&lt;sample-id&gt;*</code>: A MMseqs2 database of the input sequences (contigs + blast hits).</li>
</ul>
</li>
</ul>
</details>
<h4 id="vrhyme">vRhyme</h4>
<p><a href="https://github.com/AnantharamanLab/vRhyme">vRhyme</a> is a multi-functional tool for binning virus genomes from metagenomes. vRhyme functions by utilizing coverage variance comparisons and supervised machine learning classification of sequence features to construct viral metagenome-assembled genomes (vMAGs).</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/vrhyme</code><ul>
<li><code>&lt;sample-id&gt;/vRhyme_best_bins.#.membership.tsv</code>: scaffold membership of best bins.</li>
<li><code>&lt;sample-id&gt;/vRhyme_best_bins.#.summary.tsv</code>: summary stats of best bins.</li>
</ul>
</li>
</ul>
</details>
<h4 id="mash">Mash</h4>
<p><a href="https://github.com/marbl/Mash">Mash</a> calculates the distance between two sequences based on the jaccard distance. The Mash distance can be quickly computed from the size-reduced sketches alone, yet produces a result that strongly correlates with alignment-based measures such as the Average Nucleotide Identity (ANI).</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/mash</code><ul>
<li><code>&lt;sample-id&gt;/dist/*.tsv</code>: A distance matrix of the genomes with ANI.</li>
<li><code>&lt;sample-id&gt;/cluster/*.tsv</code>: A table where the first column represents the contig/genome and the second column it's corresponding cluster.</li>
<li><code>&lt;sample-id&gt;/visual/*.png</code>: A visualization of the network.</li>
</ul>
</li>
</ul>
</details>
<p>The network of a triple segmented Hazara virus looks like this, each node represents a contig colored on cluster. The edge represents that the ANI is higher than the specified <code>--identity_threshold</code>.</p>
<p><img alt="mash HAZV example image" class="center" src="../images/mash_HAZV_example.png" /></p>
<div class="admonition info">
<p class="admonition-title">What are those names?</p>
<p>Most assemblers tend to give each contig name a specific prefix. For example,</p>
<ul>
<li>Trinity: <code>'TRINITY_...'</code></li>
<li>SPAdes: <code>'NODE_...'</code></li>
<li>Megahit: <code>'k\d{3}_...'</code></li>
</ul>
<p>Based on these prefixes viralgenie separates external references from denovo contigs. If any assemblers are added, consider specifying a specific regex for <code>--assembler_patterns</code>.</p>
</div>
<h3 id="minimap2">Minimap2</h3>
<p><a href="https://github.com/lh3/minimap2">Minimap2</a> is a versatile sequence alignment program that aligns larger DNA or mRNA sequences against a large reference database.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/scaffolding/&lt;sample-id&gt;/minimap</code><ul>
<li><code>&lt;sample-id&gt;_cl#.bam</code>: A BAM file containing the alignment of contigs to the centroid.</li>
<li><code>&lt;sample-id&gt;_cl#.mmi</code>: The centroid index file.</li>
</ul>
</li>
</ul>
</details>
<blockquote>
<p>By default, viralgenie will not provide the minimap output files. The intermediate files can be saved by specifying <code>--save_intermediate_polishing</code>.</p>
</blockquote>
<h3 id="ivar-contig-consensus">iVar contig consensus</h3>
<p><a href="https://andersen-lab.github.io/ivar/html/manualpage.html#autotoc_md19">iVar</a> is a computational method for calling consensus sequences from viral populations.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>polishing/scaffolding/&lt;sample-id&gt;</code><ul>
<li><code>&lt;sample-id&gt;_cl#_consensus.fa</code>: A fasta file containing the consensus sequence of the cluster.</li>
<li><code>&lt;sample-id&gt;_cl#_consensus.mpileup</code>: A mpileup file containing depth at each position of the consensus sequence.</li>
<li><code>hybrid-&lt;sample-id&gt;_cl#_consensus.fa</code>: A fasta file containing the hybrid consensus sequence of the cluster and the reference.</li>
<li><code>/visualised/</code><ul>
<li><code>*.png</code>: A visualization of the consensus sequence displaying which regions came from the reference and which from the contigs.</li>
<li><code>*.txt</code>: The alignment of the reference to the consensus sequence written as a blast alignment.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<blockquote>
<p>By default, viralgenie will not provide the iVar output files. The intermediate files can be saved by specifying <code>--save_intermediate_polishing</code>.</p>
</blockquote>
<p>A visualization is made to show which regions came from the external reference (red) and which from the denovo contigs (green). For example,</p>
<p><img alt="iVar visualization example" class="center" src="../images/hybrid_consensus_scaffolding.png" /></p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The hybrid consensus is generated by mapping the contigs to the reference and then calling the consensus sequence. This is done to fill in the gaps in the contigs with the reference sequence, if there are no positions with 0 coverage there will not be a hybrid consensus and the output from iVar will be used.</p>
</div>
<h2 id="variant-calling-iterative-refinement">Variant Calling &amp; Iterative Refinement</h2>
<p>The results from variant calling, resulting from the <a href="../workflow/variant_and_refinement/">mapping constraints</a> &amp; the final round of polishing are stored in the directory <code>variant_calling/</code>.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>Mapping constraints are combined with the specified samples, here, the identifier of the mapping constraint combined with the sample identifier. All results will have a new prefix which is <code>&lt;sample-id&gt;_&lt;mapping_constraint_id&gt;-CONSTRAINT</code>.</p>
</div>
<p>The results from the iterations are stored with the same structure as the final round of polishing in the <code>assembly/polishing/iterations/it#</code> directory.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>To be able to make a distinction between the output files of the iterations, viralgenie follows a schema where it starts from <code>singletons</code> or a <code>consensus</code> goes through the iterations and ends with the <code>variant-calling</code>. The output files will have the following structure:
<pre class="mermaid"><code>graph LR
    F[singleton] --&gt; B[Iteration 1: 'it1']
    A[consensus] --&gt; B[Iteration 1: 'it1']
    B --&gt; C[Iteration 2: 'it2']
    C --&gt; D[...]
    D --&gt; E[Variant-calling: 'itvariant-calling']</code></pre></p>
<p>The prefix of the sample is combined with the previous state of the sample. For example, in the first iteration (directory <code>iterations/it1</code>), reads will be mapped to the <a href="../workflow/assembly_polishing/">reference-assisted de novo consensus sequence</a> (<em>ie</em> <code>consensus</code>) and so the output file will be <code>assembly/polishing/iterations/it1/bwamem2/bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_consensus.bam</code>.</p>
</div>
<h3 id="reference-selection">Reference selection</h3>
<p>The reference selection is done using <a href="https://mash.readthedocs.io/en/latest/index.html"><code>mash</code></a> tool. Here the reference file is sketched (<code>variants/mapping-info/mash/sketch</code>) and compared to the reads (<code>variants/mapping-info/mash/screen</code>) where the reference with the highest estimated average nucleotide identity (ANI) and shared hashes is selected (<code>variants/mapping-info/mash/select-ref</code>).</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>variants/mapping-info/mash</code><ul>
<li><code>sketch/&lt;sample-id&gt;_&lt;constraint-id&gt;.msh</code>: The sketch file of the reads.</li>
<li><code>screen/&lt;sample-id&gt;_&lt;constraint-id&gt;.screen</code>: The tab results file of the comparisons between references and reads.</li>
<li><code>select-ref/&lt;sample-id&gt;_&lt;constraint-id&gt;.json</code>: The reference with the highest estimated ANI and shared hashes.</li>
</ul>
</li>
</ul>
</details>
<details class="- info">
<summary>Column names: mash-screen</summary>
<ul>
<li>identity</li>
<li>shared-hashes</li>
<li>median-multiplicity</li>
<li>p-value</li>
<li>query-ID</li>
<li>query-comment</li>
</ul>
</details>
<h3 id="read-mapping">Read mapping</h3>
<p>The mapping results are stored in the directory <code>variants/mapping-info/</code> or in the iterations directory <code>assembly/polishing/iterations/it#</code>.</p>
<p>If bowtie is used, the output from the raw mapping results (in addition to the results after deduplication) are included in the multiqc report.</p>
<p><img alt="MultiQC - Bowtie2 alignment score plot" src="../images/mqc_bowtie2_plot.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/</code><ul>
<li><code>bwamem2/</code><ul>
<li><code>index/&lt;sample-id&gt;_&lt;constraint-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
</ul>
</li>
<li><code>bwamem/</code><ul>
<li><code>index/&lt;sample-id&gt;_&lt;constraint-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
</ul>
</li>
<li><code>bowtie2/</code><ul>
<li><code>build/&lt;sample-id&gt;_&lt;constraint-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
<li><code>log/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.log</code>: A log file of the bowtie2 run.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/</code><ul>
<li><code>bwamem2/</code><ul>
<li><code>index/&lt;sample-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
</ul>
</li>
<li><code>bwamem/</code><ul>
<li><code>index/&lt;sample-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
</ul>
</li>
<li><code>bowtie2/</code><ul>
<li><code>build/&lt;sample-id&gt;/*</code>: The index files of the consensus.</li>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>unmapped/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.fastq.gz</code>: A fastq file containing the unmapped reads.</li>
<li><code>log/&lt;sample-id&gt;_cl#_it#.log</code>: A log file of the bowtie2 run.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<h3 id="deduplication">Deduplication</h3>
<p>To accommodate for PCR duplicates, the reads are deduplicated. The deduplication results are stored in the directory <code>variants/mapping-info/deduplicate/</code> or in the iterations directory <code>assembly/polishing/iterations/it#/deduplicate</code>.</p>
<p>Deduplication results are also visualized within the MultiQC report.</p>
<h4 id="umi-tools">UMI-tools</h4>
<p><a href="https://umi-tools.readthedocs.io/en/latest/QUICK_START.html"><code>UMI-tools</code></a> is a set of tools for handling Unique Molecular Identifiers (UMIs) in NGS data. The deduplication is done by the <code>dedup</code> tool.</p>
<p><strong>Number of deduplicated reads</strong>:
<img alt="MultiQC - UMI-tools deduplication plot" class="center" src="../images/mqc_umitools_deduplication_barplot.png" /></p>
<p><strong>Summary statistics</strong>:
<img alt="MultiQC - UMI-tools violin plot" class="center" src="../images/mqc_umitools_stats_violin.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/deduplicate/</code><ul>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.umi_deduplicated.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>log/</code><ul>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.umi_deduplicated.log</code>: A log file of the UMI-tools run.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.umi_deduplicated_edit_distance.tsv</code>: Reports the (binned) average edit distance between the UMIs at each position.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.umi_deduplicated_per_umi.tsv</code>: UMI-level summary statistics.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.umi_deduplicated_per_umi_per_position.tsv</code>: Tabulates the counts for unique combinations of UMI and position.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/deduplicate</code><ul>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.umi_deduplicated.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>log/</code><ul>
<li><code>&lt;sample-id&gt;_cl#_it#.umi_deduplicated.log</code>: A log file of the UMI-tools run.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.umi_deduplicated_edit_distance.tsv</code>: Reports the (binned) average edit distance between the UMIs at each position.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.umi_deduplicated_per_umi.tsv</code>: UMI-level summary statistics.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.umi_deduplicated_per_umi_per_position.tsv</code>: Tabulates the counts for unique combinations of UMI and position.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<h4 id="picard-mark-duplicates">Picard - Mark Duplicates</h4>
<p><a href="https://broadinstitute.github.io/picard/"><code>Picard</code></a> is a set of command line tools for manipulating high-throughput sequencing data and formats such as SAM/BAM/CRAM and VCF. The deduplication is done by the <code>MarkDuplicates</code> tool.</p>
<p><img alt="MultiQC - Picard deduplication plot" src="../images/mqc_picard_deduplication.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/deduplicate/</code><ul>
<li><code>picard/</code><ul>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>log/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.dedup.MarkDuplicates.metrics.txt</code>: Deduplication metrics from Picard.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/deduplicate</code><ul>
<li><code>picard/</code><ul>
<li><code>bam/&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.bam</code>: A BAM file containing the alignment of contigs to the consensus.</li>
<li><code>log/&lt;sample-id&gt;_cl#_it#.dedup.MarkDuplicates.metrics.txt</code>: Deduplication metrics from Picard.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<h3 id="mapping-statistics">Mapping statistics</h3>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>If <code>--deduplicate</code> is set to <code>true</code> [default], all metrics will be calculated on the deduplicated bam file.</p>
</div>
<h4 id="samtools">Samtools</h4>
<p><a href="http://www.htslib.org/">Samtools</a> is a suite of programs for interacting with high-throughput sequencing data. We use samtools in this pipeline to obtain mapping statistics from three tools: <code>flagstat</code>, <code>idxstats</code> and <code>stats</code>.</p>
<p><img alt="MultiQC - Samtools stats statistics" class="center" src="../images/mqc_samtools_alignment_plot.png" />
<img alt="MultiQC - Samtools flagstat statistics" class="center" src="../images/mqc_samtools-flagstat-dp.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/metrics</code><ul>
<li><code>flagstat/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.flagstat</code>: A text file containing the flagstat output.</li>
<li><code>idxstats/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.idxstats</code>: A text file containing the idxstats output.</li>
<li><code>stats/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.stats</code>: A text file containing the stats output.</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/metrics</code><ul>
<li><code>flagstat/&lt;sample-id&gt;_cl#_it#.flagstat</code>: A text file containing the flagstat output.</li>
<li><code>idxstats/&lt;sample-id&gt;_cl#_it#.idxstats</code>: A text file containing the idxstats output.</li>
<li><code>stats/&lt;sample-id&gt;_cl#_it#.stats</code>: A text file containing the stats output.</li>
</ul>
</li>
</ul>
</details>
<h4 id="picard-collect-multiple-metrics">Picard - Collect Multiple Metrics</h4>
<p><a href="https://broadinstitute.github.io/picard/"><code>Picard</code></a> is a set of command line tools for manipulating high-throughput sequencing data. We use picard-tools in this pipeline to obtain mapping and coverage metrics.</p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/metrics/picard</code><ul>
<li><code>*.CollectMultipleMetrics.*</code>: Alignment QC files from picard CollectMultipleMetrics in <code>*_metrics</code> textual format.</li>
<li><code>*.pdf</code> plots for metrics obtained from CollectMultipleMetrics.</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/metrics/picard</code><ul>
<li><code>*.CollectMultipleMetrics.*</code>: Alignment QC files from picard CollectMultipleMetrics in <code>*_metrics</code> textual format.</li>
<li><code>*.pdf</code> plots for metrics obtained from CollectMultipleMetrics.</li>
</ul>
</li>
</ul>
</details>
<h4 id="custom-mpileup-like-file">Custom - mpileup like file</h4>
<p>To facilitate the intra host analysis, a mpileup like file is generated. This file contains the depth of every nucleotide at each position of the reference as well as the shannon entropy and a weighted shannon entropy based on the following formulae.</p>
<ul>
<li>Shannon entropy: $$ H = -\sum_{i=1}^4 p_i \ln p_i $$</li>
<li>Weighted Shannon entropy: $$ w(H) = \frac{N}{N+k} \cdot H $$</li>
</ul>
<p>Where <span class="arithmatex">\(N\)</span> is the total bases at a position, <span class="arithmatex">\(k\)</span> is the pseudocount (default 50), and <span class="arithmatex">\(p_i\)</span> is the frequency of the nucleotide <span class="arithmatex">\(i\)</span>.</p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/custom-vcf/&lt;sample-id&gt;</code><ul>
<li><code>*.tsv</code>: A custom tsv file containing the depth of every nucleotide at each position of the reference.</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/custom-vcf/&lt;sample-id&gt;</code><ul>
<li><code>*.tsv</code>: A custom tsv file containing the depth of every nucleotide at each position of the reference.</li>
</ul>
</li>
</ul>
</details>
<h4 id="mosdepth-coverage">Mosdepth - Coverage</h4>
<p><a href="https://github.com/brentp/mosdepth">mosdepth</a> is a fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing. mosdepth is used in this pipeline to obtain genome-wide coverage values in 200bp windows. The results are rendered in MultiQC (genome-wide coverage).</p>
<p><img alt="MultiQC - Mosdepth cumulative coverage plot" class="center" src="../images/mqc_mosdepth-cumcoverage-dist-id.png" />
<img alt="MultiQC - Mosdepth coverage plot" class="center" src="../images/mqc_mosdepth-coverage-per-contig-single.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/mapping-info/metrics/mosdepth</code><ul>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.per-base.bed.gz</code>: A bed file containing the coverage values in 200bp windows.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.per-base.bed.gz.csi</code>: Indexed bed file.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.mosdepth.summary.txt</code>: Summary metrics including mean, min and max coverage values.</li>
<li><code>&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.mosdepth.global.dist.txt</code>: A cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value.</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<ul>
<li><code>assembly/polishing/iterations/it#/metrics/mosdepth</code><ul>
<li><code>&lt;sample-id&gt;_cl#_it#.per-base.bed.gz</code>: A bed file containing the coverage values in 200bp windows.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.per-base.bed.gz.csi</code>: Indexed bed file.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.mosdepth.summary.txt</code>: Summary metrics including mean, min and max coverage values.</li>
<li><code>&lt;sample-id&gt;_cl#_it#.mosdepth.global.dist.txt</code>: A cumulative distribution indicating the proportion of total bases that were covered for at least a given coverage value.</li>
</ul>
</li>
</ul>
</details>
<h3 id="variant-calling-filtering">Variant calling &amp; filtering</h3>
<p>Variant calling is done with <a href="http://samtools.github.io/bcftools/bcftools.html"><code>BCFTools mpileup</code></a> or <a href="https://andersen-lab.github.io/ivar/html/manualpage.html"><code>iVar</code></a>, the filtering with <code>BCFtools filter</code>.</p>
<p>Variant files are visualized in the MultiQC report.</p>
<p><img alt="MultiQC - BCFTools variant calling plot" class="center" src="../images/mqc_bcftools-stats-subtypes.png" /></p>
<details class="- abstract">
<summary>Output files - variants</summary>
<ul>
<li><code>variants/variant_calling</code><ul>
<li><code>bcftools/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.vcf.gz</code>: A VCF file containing the variant calls.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.norm.vcf.gz</code>: A compressed VCF file where multiallelic sites are split up into biallelic records and SNPs and indels should be merged into a single record.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.filtered.vcf.gz</code>: A compressed VCF file containing the filtered variants.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.vcf.gz.tbi</code>: An index file for the compressed VCF file.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT_stats.txt</code>: A text file stats which is suitable for machine processing and can be plotted using plot-vcfstats.</li>
</ul>
</li>
<li><code>ivar/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.ivar.tsv</code>: A tabular file containing the variant calls.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.ivar.vcf</code>: A VCF file containing the variant calls.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.ivar.variant_counts.log</code>: A summary file containing the number of indels and SNPs.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT.filtered.vcf.gz</code>: A compressed VCF file containing the variant calls.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;constraint-id&gt;-CONSTRAINT_stats.txt</code>: A text file stats which is suitable for machine processing and can be plotted using plot-vcfstats.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<details class="- abstract">
<summary>Output files - iterations</summary>
<div class="language-text highlight"><pre><span></span><code>- `assembly/polishing/iterations/it#/variants/variant_calling`
    - `bcftools/`
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.vcf.gz`: A VCF file containing the variant calls.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.norm.vcf.gz`: A compressed VCF file where multiallelic sites are split up into biallelic records and SNPs and indels should be merged into a single record.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.filtered.vcf.gz`: A compressed VCF file containing the filtered variants.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.vcf.gz.tbi`: An index file for the compressed VCF file.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.stats.txt`: A text file stats which is suitable for machine processing and can be plotted using plot-vcfstats.
    - `ivar/`
        - `&lt;sample-id&gt;_cl#_it#.ivar.tsv`: A tabular file containing the variant calls.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.ivar.vcf`: A VCF file containing the variant calls.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.ivar.variant_counts.log`: A summary file containing the number of indels and SNPs.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.filtered.vcf.gz`: A compressed VCF file containing the variant calls.
        - `&lt;sample-id&gt;/&lt;sample-id&gt;_cl#_it#.stats.txt`: A text file stats which is suitable for machine processing and can be plotted using plot-vcfstats.
</code></pre></div>
</details>
<h3 id="consensus-generation">Consensus generation</h3>
<p>The consensus sequences are generated by <a href="http://samtools.github.io/bcftools/bcftools.html"><code>BCFTools</code></a> or <a href="https://andersen-lab.github.io/ivar/html/manualpage.html"><code>iVar</code></a>. The consensus sequences are stored in the directory <code>consensus/</code> or in the iterations directory <code>assembly/polishing/iterations/it#/consensus</code>.</p>
<p><code>BCFtools</code> will use the filtered variants file whereas, <code>iVar</code> will redetermine the variants to collapse in the consensus using their own workflow, read more about their differences in the <a href="../workflow/variant_and_refinement/#4-consensus-calling">consensus calling section</a>.</p>
<details class="- abstract">
<summary>Output files - iterations &amp; variants</summary>
<ul>
<li><code>consensus</code><ul>
<li><code>seq/&lt;it# | scaffold_consensus | variant-calling | constraint&gt;/</code><ul>
<li><code>&lt;sample-id&gt;/*.fasta</code>: A fasta file containing the consensus sequence.</li>
</ul>
</li>
<li><code>mask/&lt;it# | variant-calling | constraint&gt;</code><ul>
<li><code>&lt;sample-id&gt;/*.qual.txt</code>: A log file of the consensus run containing statistics. [<code>iVar</code> only]</li>
<li><code>&lt;sample-id&gt;/*.bed</code>: A bed file containing the masked regions. [<code>BCFtools</code> only]</li>
<li><code>&lt;sample-id&gt;/*.mpileup</code>: A mpileup file containing information on the depth and the quality of each aligned base.</li>
</ul>
</li>
</ul>
</li>
</ul>
</details>
<h2 id="consensus-quality-control">Consensus Quality control</h2>
<p>Consensus quality control is done with multiple tools, the results are stored in the directory <code>consensus/quality_control/</code>.</p>
<h3 id="quast">Quast</h3>
<p><a href="http://quast.sourceforge.net/">QUAST</a> is a quality assessment tool for genome assemblies. It calculates various metrics such as N50, L50, number of contigs, number of mismatches, number of indels, etc.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>consensus/quality_control/quast/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;iteration&gt;/&lt;sample-id&gt;_&lt;cl# | constraint-id&gt;.tsv</code>: A tabular file containing the QUAST report.</li>
</ul>
</li>
</ul>
<blockquote>
<p>If no iterative refinement was run, the output will be in the <code>consensus/quality_control/quast/&lt;sample-id&gt;/constraint</code> directory.</p>
</blockquote>
</details>
<h3 id="checkv">CheckV</h3>
<p><a href="https://bitbucket.org/berkeleylab/checkv/src/master/"><code>CheckV</code></a> is a tool for assessing the quality of viral genomes recovered from metagenomes. It calculates various metrics such as the number of viral genes, the number of viral contigs, the number of viral genomes, etc.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>consensus/quality_control/checkv/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;cl# | constraint-id&gt;/quality_summary.tsv</code>: A tabular file that integrates the results from the three main modules of checkv and should be the main output referred to.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;cl# | constraint-id&gt;/completeness.tsv</code>: A detailed overview of how completeness was estimated.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;cl# | constraint-id&gt;/contamination.tsv</code>: A detailed overview of how contamination was estimated.</li>
<li><code>&lt;sample-id&gt;/&lt;sample-id&gt;_&lt;cl# | constraint-id&gt;/complete_genomes.tsv</code>: A detailed overview of putative genomes identified.</li>
</ul>
</li>
</ul>
</details>
<h3 id="blastn">BLASTn</h3>
<p><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> is a tool for comparing primary biological sequence information. The output from the BLAST run is stored in the directory <code>consensus/quality_control/blast/</code>. Final consensus genomes are searched against the <code>--reference_pool</code>.</p>
<details class="- info">
<summary>Column names</summary>
<ul>
<li>qseqid</li>
<li>sseqid</li>
<li>stitle</li>
<li>pident</li>
<li>qlen</li>
<li>slen</li>
<li>length</li>
<li>mismatch</li>
<li>gapopen</li>
<li>qstart</li>
<li>qend</li>
<li>sstart</li>
<li>send</li>
<li>evalue</li>
<li>bitscore</li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Modifying blast columns</p>
<p>Modifying these columns can be done through a custom config file and by updating <code>bin/utils/constant_variables.py</code>.</p>
</div>
</details>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>consensus/quality_control/blast/</code> A tabular file containing the BLAST report of all intermediate &amp; final results.</li>
</ul>
</details>
<h3 id="mmseqs-search-annotation">MMseqs-search (annotation)</h3>
<p><a href="https://github.com/soedinglab/MMseqs2/wiki#searching">MMseqs-search</a> is an ultra-fast and sensitive search tool for protein and nucleotide databases. Viralgenie uses MMseqs to search the consensus genomes in an annotated database, like <a href="https://virosaurus.vital-it.ch/">Virousarus</a> (see also <a href="../customisation/databases/#annotation-sequences">defining your own custom annotation database</a>), and uses the annotation data of the best hit to assign the consensus genome a species name, segment name, expected host, and any other metadata that is embedded within the database.</p>
<details class="- info">
<summary>Column names</summary>
<ul>
<li>qseqid</li>
<li>sseqid</li>
<li>stitle</li>
<li>pident</li>
<li>qlen</li>
<li>slen</li>
<li>length</li>
<li>mismatch</li>
<li>gapopen</li>
<li>qstart</li>
<li>qend</li>
<li>sstart</li>
<li>send</li>
<li>evalue</li>
<li>bitscore</li>
</ul>
<div class="admonition tip">
<p class="admonition-title">Modifying mmseqs columns</p>
<p>Modifying these columns can be done through a custom config file and by updating <code>bin/utils/constant_variables.py</code>.</p>
</div>
</details>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>consensus/quality_control/mmseqs-search/all_genomes_annotation.hits.tsv</code>: A tabular file containing the MMseqs-search hits, all genomes are combined to reduce the number of jobs.</li>
</ul>
</details>
<h3 id="mafft">MAFFT</h3>
<p><a href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> is a multiple sequence alignment program for amino acid or nucleotide sequences. The output from the MAFFT run is stored in the directory <code>consensus/quality_control/mafft/</code>.</p>
<p>It is used to align the following genomic data:
- The final consensus genome
- The identified reference genome from <code>--reference_pool</code>
- The denovo contigs from each assembler (that constituted the final consensus genome)
- Each consensus genome from the iterative refinement steps.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>consensus/quality_control/mafft/</code><ul>
<li><code>&lt;sample-id&gt;/&lt;sample_id&gt;_cl#_iterations.fas</code>: A fasta file containing a multiple sequence alignment of only the iterations.</li>
<li><code>&lt;sample-id&gt;/&lt;sample_id&gt;_cl#_aligned.fas</code>: A fasta file containing a multiple sequence alignment of the denovo contigs, the reference from reference_pool and the consensus from iterations.</li>
</ul>
</li>
</ul>
</details>
<p>Alignment can then be opened with MSA viewer, for example Jalview</p>
<p><img alt="MSA alignment jalview" class="center" src="../images/mafft_example_alignment_jalview.png" /></p>
<h2 id="multiqc">MultiQC</h2>
<p><a href="http://multiqc.info">MultiQC</a> is a visualization tool that generates a single HTML report summarizing all samples in your project. Most of the pipeline QC results are visualized in the report and further statistics are available in the report data directory.</p>
<p>Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <a href="http://multiqc.info">http://multiqc.info</a>.</p>
<p>Furthermore, viralgenie runs MultiQC 2 times, as it uses the output from multiqc to create multiple summary tables of the consensus genomes and their iterations.</p>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>multiqc/</code><ul>
<li><code>overview-tables/</code>: a directory with a set of commented TSV (comments taken from <code>--multiqc_comment_headers</code>) that summarize aspects of the pipeline runs.</li>
<li><code>multiqc_report.html</code>: a standalone HTML file that can be viewed in your web browser.</li>
<li><code>multiqc_data/</code>: directory containing parsed statistics from the different tools used in the pipeline.</li>
<li><code>multiqc_dataprep/</code>: preparation files for the generated custom tables.</li>
<li><code>multiqc_plots/</code>: directory containing static images from the report in various formats.</li>
</ul>
</li>
</ul>
</details>
<h2 id="pipeline-information">Pipeline information</h2>
<details class="- abstract">
<summary>Output files</summary>
<ul>
<li><code>pipeline_info/</code><ul>
<li>Reports generated by Nextflow: <code>execution_report.html</code>, <code>execution_timeline.html</code>, <code>execution_trace.txt</code> and <code>pipeline_dag.dot</code>/<code>pipeline_dag.svg</code>.</li>
<li>Reports generated by the pipeline: <code>pipeline_report.html</code>, <code>pipeline_report.txt</code> and <code>software_versions.yml</code>. The <code>pipeline_report*</code> files will only be present if the <code>--email</code> / <code>--email_on_fail</code> parameter's are used when running the pipeline.</li>
<li>Reformatted samplesheet files used as input to the pipeline: <code>samplesheet.valid.csv</code>.</li>
<li>Parameters used by the pipeline run: <code>params.json</code>.</li>
</ul>
</li>
</ul>
</details>
<p><a href="https://www.nextflow.io/docs/latest/tracing.html">Nextflow</a> provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
````</p>







  
  



  


  


  <form class="md-feedback" name="feedback" hidden>
    <fieldset>
      <legend class="md-feedback__title">
        Was this page helpful?
      </legend>
      <div class="md-feedback__inner">
        <div class="md-feedback__list">
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page was helpful" data-md-value="1">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12 21.35-1.45-1.32C5.4 15.36 2 12.27 2 8.5 2 5.41 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.08C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.41 22 8.5c0 3.77-3.4 6.86-8.55 11.53z"/></svg>
            </button>
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page could be improved" data-md-value="0">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12 21.35-1.45-1.32C5.4 15.36 2 12.27 2 8.5 2 5.41 4.42 3 7.5 3c.67 0 1.32.12 1.94.33L13 9.35l-4 5zM16.5 3C19.58 3 22 5.41 22 8.5c0 3.77-3.4 6.86-8.55 11.53L12 21.35l-1-7 4.5-5-2.65-5.08C13.87 3.47 15.17 3 16.5 3"/></svg>
            </button>
          
        </div>
        <div class="md-feedback__note">
          
            <div data-md-value="1" hidden>
              
              
                
              
              Thanks for your feedback, I like you too!
            </div>
          
            <div data-md-value="0" hidden>
              
              
                
              
              Thanks for your feedback! Help us improve this page by submitting an <a href="https://github.com/Joon-Klaps/viralgenie/issues/new/?labels=website%2Cdocumentation&template=website_feedback.yml&title=[Website feedback]+Output" target="_blank" rel="noopener">issue</a>.
            </div>
          
        </div>
      </div>
    </fieldset>
  </form>


                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
        
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../parameters/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Parameters">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Parameters
              </div>
            </div>
          </a>
        
        
          
          <a href="../workflow/" class="md-footer__link md-footer__link--next" aria-label="Next: Workflow overview">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Workflow overview
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://twitter.com/LemeyLab" target="_blank" rel="noopener" title="twitter.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8l164.9-188.5L26.8 48h145.6l100.5 132.9zm-24.8 373.8h39.1L151.1 88h-42z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/Joon-Klaps" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            

  
    
  


  
    
  



  


<h4>Cookie Consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our documentation better.</p>
<input class="md-toggle" type="checkbox" id="__settings" >
<div class="md-consent__settings">
  <ul class="task-list">
    
      
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="analytics" checked>
          <span class="task-list-indicator"></span>
          Google Analytics
        </label>
      </li>
    
      
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="github" checked>
          <span class="task-list-indicator"></span>
          GitHub
        </label>
      </li>
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Accept</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Manage settings</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout((function(){document.querySelector("[data-md-component=consent]").hidden=!1}),250);var form=document.forms.consent;for(var action of["submit","reset"])form.addEventListener(action,(function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map((function(e){return[e,!0]})))),location.hash="",location.reload()}))</script>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["announce.dismiss", "content.code.annotate", "content.code.copy", "header.autohide", "navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.indexes", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.share", "search.suggest", "toc.follow"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../javascripts/tablesort.js"></script>
      
        <script src="../javascripts/katex.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/katex.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>