Changes in module 'nf-core/cdhit/cdhitest'
--- modules/nf-core/cdhit/cdhitest/main.nf
+++ modules/nf-core/cdhit/cdhitest/main.nf
@@ -22,6 +22,8 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     def suffix = task.ext.suffix ?: "${sequences}" ==~ /(.*f[astn]*a(.gz)?$)/ ? "fa" : "fq"
+    def pident = meta.pident ? + meta.pident : params.identity_threshold
+    def score = pident < 0.8 ? '-c 0.8' : '-c ' + pident
 
     def avail_mem = 3072
     if (!task.memory) {
@@ -32,8 +34,9 @@
     """
     cd-hit-est \\
         $args \\
+        ${score} \\
         -i ${sequences} \\
-        -o ${meta.id}.${suffix} \\
+        -o ${prefix}.${suffix} \\
         -M $avail_mem \\
         -T $task.cpus
 

************************************************************
