Changes in module 'nf-core/multiqc'
--- modules/nf-core/multiqc/environment.yml
+++ modules/nf-core/multiqc/environment.yml
@@ -4,4 +4,4 @@
   - bioconda
   - defaults
 dependencies:
-  - bioconda::multiqc=1.21
+  - bioconda::multiqc=1.22.1

--- modules/nf-core/multiqc/main.nf
+++ modules/nf-core/multiqc/main.nf
@@ -3,8 +3,8 @@
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/multiqc:1.21--pyhdfd78af_0' :
-        'biocontainers/multiqc:1.21--pyhdfd78af_0' }"
+        'oras://community.wave.seqera.io/library/multiqc:1.22.1--ac0a91c1ae1c160c' :
+        'community.wave.seqera.io/library/multiqc:1.22.1--4886de6095538010' }"
 
     input:
     path  multiqc_files, stageAs: "?/*"
@@ -13,7 +13,7 @@
     path(multiqc_logo)
 
     output:
-    path "*multiqc_report.html", emit: report
+    path "*multiqc_report.html", optional:true, emit: report
     path "*_data"              , emit: data
     path "*_plots"             , optional:true, emit: plots
     path "versions.yml"        , emit: versions

************************************************************
