# Citations

## [Viralgenie](https://github.com/Joon-Klaps/viralgenie)

!!! warning
    Viralgenie is currently not Published. Please cite as:

    - Klaps J, Lemey P, Kafetzopoulou L. Viralgenie: A metagenomics analysis pipeline for eukaryotic viruses. __Github__ https://github.com/Joon-Klaps/viralgenie

## [nf-core](https://pubmed.ncbi.nlm.nih.gov/32055031/)

> Ewels PA, Peltzer A, Fillinger S, Patel H, Alneberg J, Wilm A, Garcia MU, Di Tommaso P, Nahnsen S. The nf-core framework for community-curated bioinformatics pipelines. Nat Biotechnol. 2020 Mar;38(3):276-278. doi: 10.1038/s41587-020-0439-x. PubMed PMID: 32055031.

## [Nextflow](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

- [Bbduk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/)

    > Bushnell B. (2022) BBMap, URL: http://sourceforge.net/projects/bbmap/

- [BCFtools](https://pubmed.ncbi.nlm.nih.gov/33590861/)

    > Danecek, Petr et al. “Twelve years of SAMtools and BCFtools.” GigaScience vol. 10,2 (2021): giab008. doi:10.1093/gigascience/giab008

- [blast](https://pubmed.ncbi.nlm.nih.gov/20003500/)

    >Camacho, Christiam et al. “BLAST+: architecture and applications.” BMC bioinformatics vol. 10 421. 15 Dec. 2009, doi:10.1186/1471-2105-10-421

- [Bowtie2](https://bowtie-bio.sourceforge.net/bowtie2/index.shtml)

    > Langmead, Ben, and Steven L Salzberg. “Fast gapped-read alignment with Bowtie 2.” Nature methods vol. 9,4 357-9. 4 Mar. 2012, doi:10.1038/nmeth.1923

- [BWA-MEM](https://github.com/lh3/bwa)

    > Li H. (2013) Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv:1303.3997v2.

- [BWA-MEM2](https://github.com/bwa-mem2/bwa-mem2)

    > M. Vasimuddin, S. Misra, H. Li and S. Aluru, "Efficient Architecture-Aware Acceleration of BWA-MEM for Multicore Systems," 2019 IEEE International Parallel and Distributed Processing Symposium (IPDPS), Rio de Janeiro, Brazil, 2019, pp. 314-324, doi: 10.1109/IPDPS.2019.00041.

- [cdhit](https://pubmed.ncbi.nlm.nih.gov/23060610/)

    > Fu, Limin et al. “CD-HIT: accelerated for clustering the next-generation sequencing data.” Bioinformatics (Oxford, England) vol. 28,23 (2012): 3150-2. doi:10.1093/bioinformatics/bts565

- [checkv](https://pubmed.ncbi.nlm.nih.gov/33349699/)

    > Nayfach, Stephen et al. “CheckV assesses the quality and completeness of metagenome-assembled viral genomes.” Nature biotechnology vol. 39,5 (2021): 578-585. doi:10.1038/s41587-020-00774-7

- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

    > Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data [Online].

- [fastp](https://github.com/OpenGene/fastp)

    > Chen, Shifu et al. “fastp: an ultra-fast all-in-one FASTQ preprocessor.” Bioinformatics (Oxford, England) vol. 34,17 (2018): i884-i890. doi:10.1093/bioinformatics/bty560

- [HUMID](https://humid.readthedocs.io/en/latest/usage.html)
<!-- TODO  CONTINUE HERE -->
    > __Github__ https://github.com/jfjlaros/HUMID

- [iVar](https://www.ncbi.nlm.nih.gov/pubmed/30621750/)

    > Grubaugh, Nathan D et al. “An amplicon-based sequencing framework for accurately measuring intrahost virus diversity using PrimalSeq and iVar.” Genome biology vol. 20,1 8. 8 Jan. 2019, doi:10.1186/s13059-018-1618-7

- [Kaiju](https://pubmed.ncbi.nlm.nih.gov/27071849/)

    > Menzel, Peter et al. “Fast and sensitive taxonomic classification for metagenomics with Kaiju.” Nature communications vol. 7 11257. 13 Apr. 2016, doi:10.1038/ncomms11257

- [Kraken2](https://doi.org/10.1186/s13059-019-1891-0)

    > Wood, Derrick E., Jennifer Lu, and Ben Langmead. 2019. Improved Metagenomic Analysis with Kraken 2. Genome Biology 20 (1): 257. doi: 10.1186/s13059-019-1891-0.

- [leiden-algorithm](https://pubmed.ncbi.nlm.nih.gov/30914743/)

    > Traag, V A et al. “From Louvain to Leiden: guaranteeing well-connected communities.” Scientific reports vol. 9,1 5233. 26 Mar. 2019, doi:10.1038/s41598-019-41695-z

- [Mash](https://pubmed.ncbi.nlm.nih.gov/27323842/)

    > Ondov, Brian D et al. “Mash: fast genome and metagenome distance estimation using MinHash.” Genome biology vol. 17,1 132. 20 Jun. 2016, doi:10.1186/s13059-016-0997-x

- [Megahit](https://pubmed.ncbi.nlm.nih.gov/27012178/)

    > Li, Dinghua et al. “MEGAHIT v1.0: A fast and scalable metagenome assembler driven by advanced methodologies and community practices.” Methods (San Diego, Calif.) vol. 102 (2016): 3-11. doi:10.1016/j.ymeth.2016.02.020

- [Minimap2](https://pubmed.ncbi.nlm.nih.gov/29750242/)

    > Li, Heng. “Minimap2: pairwise alignment for nucleotide sequences.” Bioinformatics (Oxford, England) vol. 34,18 (2018): 3094-3100. doi:10.1093/bioinformatics/bty191

- [MMseqs2](https://pubmed.ncbi.nlm.nih.gov/29035372/)

    > Steinegger, Martin, and Johannes Söding. “MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets.” Nature biotechnology vol. 35,11 (2017): 1026-1028. doi:10.1038/nbt.3988

- [Mosdepth](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6030888/)

    > Pedersen, Brent S, and Aaron R Quinlan. “Mosdepth: quick coverage calculation for genomes and exomes.” Bioinformatics (Oxford, England) vol. 34,5 (2018): 867-868. doi:10.1093/bioinformatics/btx699

- [MultiQC](https://pubmed.ncbi.nlm.nih.gov/27312411/)

    > Ewels, Philip et al. “MultiQC: summarize analysis results for multiple tools and samples in a single report.” Bioinformatics (Oxford, England) vol. 32,19 (2016): 3047-8. doi:10.1093/bioinformatics/btw354

- [picard-tools](http://broadinstitute.github.io/picard)

- [QUAST](https://www.ncbi.nlm.nih.gov/pubmed/23422339/)

    > Gurevich, Alexey et al. “QUAST: quality assessment tool for genome assemblies.” Bioinformatics (Oxford, England) vol. 29,8 (2013): 1072-5. doi:10.1093/bioinformatics/btt086

- [SAMtools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3198575/)

    > Li H. A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. Bioinformatics. 2011 Nov 1;27(21):2987-93. doi: 10.1093/bioinformatics/btr509. Epub 2011 Sep 8. PMID: 21903627; PMCID: PMC3198575.

- [SPAdes](https://www.ncbi.nlm.nih.gov/pubmed/24093227/)

    > Bankevich, Anton et al. “SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing.” Journal of computational biology : a journal of computational molecular cell biology vol. 19,5 (2012): 455-77. doi:10.1089/cmb.2012.0021

- [Trimmomatic](https://pubmed.ncbi.nlm.nih.gov/24695404/)

    > Bolger, Anthony M et al. “Trimmomatic: a flexible trimmer for Illumina sequence data.” Bioinformatics (Oxford, England) vol. 30,15 (2014): 2114-20. doi:10.1093/bioinformatics/btu170

- [Trinity](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3571712/)

    > Haas, Brian J et al. “De novo transcript sequence reconstruction from RNA-seq using the Trinity platform for reference generation and analysis.” Nature protocols vol. 8,8 (2013): 1494-512. doi:10.1038/nprot.2013.084

- [UMI-tools](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5340976/)

    > Smith, Tom et al. “UMI-tools: modeling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy.” Genome research vol. 27,3 (2017): 491-499. doi:10.1101/gr.209601.116

- [vRhyme](https://pubmed.ncbi.nlm.nih.gov/35544285/)

    > Kieft, Kristopher et al. “vRhyme enables binning of viral genomes from metagenomes.” Nucleic acids research vol. 50,14 (2022): e83. doi:10.1093/nar/gkac341

- [VSEARCH](https://pubmed.ncbi.nlm.nih.gov/27521926/)

    > Rognes, Torbjørn et al. “VSEARCH: a versatile open source tool for metagenomics.” PeerJ vol. 4 e2584. 18 Oct. 2016, doi:10.7717/peerj.2584


## Software packaging/containerisation tools

-   [Anaconda](https://anaconda.com)

    > Anaconda Software Distribution. Computer software. Vers. 2-2.4.0. Anaconda, Nov. 2016. Web.

-   [Bioconda](https://pubmed.ncbi.nlm.nih.gov/29967506/)

    > Grüning B, Dale R, Sjödin A, Chapman BA, Rowe J, Tomkins-Tinch CH, Valieris R, Köster J; Bioconda Team. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods. 2018 Jul;15(7):475-476. doi: 10.1038/s41592-018-0046-7. PubMed PMID: 29967506.

-   [BioContainers](https://pubmed.ncbi.nlm.nih.gov/28379341/)

    > da Veiga Leprevost F, Grüning B, Aflitos SA, Röst HL, Uszkoreit J, Barsnes H, Vaudel M, Moreno P, Gatto L, Weber J, Bai M, Jimenez RC, Sachsenberg T, Pfeuffer J, Alvarez RV, Griss J, Nesvizhskii AI, Perez-Riverol Y. BioContainers: an open-source and community-driven framework for software standardization. Bioinformatics. 2017 Aug 15;33(16):2580-2582. doi: 10.1093/bioinformatics/btx192. PubMed PMID: 28379341; PubMed Central PMCID: PMC5870671.

-   [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

    > Merkel, D. (2014). Docker: lightweight linux containers for consistent development and deployment. Linux Journal, 2014(239), 2. doi: 10.5555/2600239.2600241.

-   [Singularity](https://pubmed.ncbi.nlm.nih.gov/28494014/)

    > Kurtzer GM, Sochat V, Bauer MW. Singularity: Scientific containers for mobility of compute. PLoS One. 2017 May 11;12(5):e0177459. doi: 10.1371/journal.pone.0177459. eCollection 2017. PubMed PMID: 28494014; PubMed Central PMCID: PMC5426675.
