Changes in module 'nf-core/kraken2/kraken2'
--- modules/nf-core/kraken2/kraken2/main.nf
+++ modules/nf-core/kraken2/kraken2/main.nf
@@ -26,20 +26,21 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def paired       = meta.single_end ? "" : "--paired"
+    def paired       = meta.single_end ? "" : "--paired"  
     def classified   = meta.single_end ? "${prefix}.classified.fastq"   : "${prefix}.classified#.fastq"
     def unclassified = meta.single_end ? "${prefix}.unclassified.fastq" : "${prefix}.unclassified#.fastq"
     def classified_option = save_output_fastqs ? "--classified-out ${classified}" : ""
     def unclassified_option = save_output_fastqs ? "--unclassified-out ${unclassified}" : ""
     def readclassification_option = save_reads_assignment ? "--output ${prefix}.kraken2.classifiedreads.txt" : "--output /dev/null"
     def compress_reads_command = save_output_fastqs ? "pigz -p $task.cpus *.fastq" : ""
+    def gzipped = reads[0].endsWith('.gz') ? '--gzip-compressed' : ''
 
     """
     kraken2 \\
         --db $db \\
         --threads $task.cpus \\
         --report ${prefix}.kraken2.report.txt \\
-        --gzip-compressed \\
+        $gzipped \\
         $unclassified_option \\
         $classified_option \\
         $readclassification_option \\

************************************************************
