Changes in module 'nf-core/vrhyme/vrhyme'
--- modules/nf-core/vrhyme/vrhyme/main.nf
+++ modules/nf-core/vrhyme/vrhyme/main.nf
@@ -8,8 +8,7 @@
         'biocontainers/vrhyme:1.1.0--pyhdfd78af_1' }"
 
     input:
-    tuple val(meta), path(reads)
-    tuple val(meta), path(fasta)
+    tuple val(meta), path(fasta), path(reads)
 
     output:
     tuple val(meta), path("vRhyme_best_bins_fasta/")                , emit: bins

************************************************************
