Changes in module 'nf-core/ivar/variants'
--- modules/nf-core/ivar/variants/main.nf
+++ modules/nf-core/ivar/variants/main.nf
@@ -31,8 +31,8 @@
     """
     samtools \\
         mpileup \\
+        --reference $fasta \\
         $args2 \\
-        --reference $fasta \\
         $bam \\
         $mpileup \\
         | ivar \\
@@ -45,6 +45,7 @@
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         ivar: \$(echo \$(ivar version 2>&1) | sed 's/^.*iVar version //; s/ .*\$//')
+        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
     END_VERSIONS
     """
 }

************************************************************
