
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://joon-klaps.github.io/viralgenie/latest/dev/workflow/assembly_polishing/">
      
      
        <link rel="prev" href="../metagenomic_diversity/">
      
      
        <link rel="next" href="../variant_and_refinement/">
      
      
      <link rel="icon" href="../../images/ViralGenie-nf-core-theme.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.5">
    
    
      
        <title>Assembly & Polishing - Viralgenie</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:300,300i,400,400i,700,700i%7CUbuntu+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Ubuntu";--md-code-font:"Ubuntu Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
      <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/katex.min.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-DJ9PDTBEEC"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-DJ9PDTBEEC",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-DJ9PDTBEEC",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>if("undefined"!=typeof __md_analytics){var consent=__md_get("__consent");consent&&consent.analytics&&__md_analytics()}</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#assembly-polishing" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Viralgenie" class="md-header__button md-logo" aria-label="Viralgenie" data-md-component="logo">
      
  <img src="../../images/ViralGenie-nf-core-theme.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Viralgenie
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Assembly & Polishing
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9zM20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12zm-9.15 3.96h2.3L12 9z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="custom" data-md-color-accent="custom"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/Joon-Klaps/viralgenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Joon-Klaps/viralgenie
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../usage/" class="md-tabs__link">
        
  
    
  
  Usage

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../parameters/" class="md-tabs__link">
        
  
    
  
  Parameters

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../output/" class="md-tabs__link">
        
  
    
  
  Output

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
    
  
  Workflows

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../customisation/" class="md-tabs__link">
          
  
    
  
  Customisation

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../CITATIONS/" class="md-tabs__link">
        
  
    
  
  Citations

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../CONTRIBUTING/" class="md-tabs__link">
        
  
    
  
  Contributing

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Viralgenie" class="md-nav__button md-logo" aria-label="Viralgenie" data-md-component="logo">
      
  <img src="../../images/ViralGenie-nf-core-theme.png" alt="logo">

    </a>
    Viralgenie
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Joon-Klaps/viralgenie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    Joon-Klaps/viralgenie
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Workflows
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_7" id="__nav_7_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            Workflows
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../preprocessing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Preprocessing
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../metagenomic_diversity/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Metagenomic diversity
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Assembly & Polishing
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Assembly & Polishing
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-de-novo-assembly" class="md-nav__link">
    <span class="md-ellipsis">
      1. De-novo Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-extension" class="md-nav__link">
    <span class="md-ellipsis">
      2. Extension
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-coverage-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      3. Coverage calculation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-reference-matching" class="md-nav__link">
    <span class="md-ellipsis">
      4. Reference Matching
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-taxonomy-guided-clustering" class="md-nav__link">
    <span class="md-ellipsis">
      5. Taxonomy guided Clustering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Taxonomy guided Clustering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#51-pre-clustering-using-taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      5.1 Pre-clustering using taxonomy
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#52-actual-clustering-on-nucleotide-similarity" class="md-nav__link">
    <span class="md-ellipsis">
      5.2 Actual clustering on nucleotide similarity
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-coverage-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      6. Coverage filtering
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-scaffolding" class="md-nav__link">
    <span class="md-ellipsis">
      7. Scaffolding
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-annotation-with-reference" class="md-nav__link">
    <span class="md-ellipsis">
      8. Annotation with Reference
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_and_refinement/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant analysis & Iterative refinement
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../consensus_qc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Consensus evaluation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_8" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../customisation/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Customisation
    
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_8" id="__nav_8_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            Customisation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../customisation/databases/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Databases
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../customisation/configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Configuration
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../CITATIONS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Citations
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../CONTRIBUTING/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-de-novo-assembly" class="md-nav__link">
    <span class="md-ellipsis">
      1. De-novo Assembly
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-extension" class="md-nav__link">
    <span class="md-ellipsis">
      2. Extension
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-coverage-calculation" class="md-nav__link">
    <span class="md-ellipsis">
      3. Coverage calculation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-reference-matching" class="md-nav__link">
    <span class="md-ellipsis">
      4. Reference Matching
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-taxonomy-guided-clustering" class="md-nav__link">
    <span class="md-ellipsis">
      5. Taxonomy guided Clustering
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Taxonomy guided Clustering">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#51-pre-clustering-using-taxonomy" class="md-nav__link">
    <span class="md-ellipsis">
      5.1 Pre-clustering using taxonomy
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#52-actual-clustering-on-nucleotide-similarity" class="md-nav__link">
    <span class="md-ellipsis">
      5.2 Actual clustering on nucleotide similarity
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-coverage-filtering" class="md-nav__link">
    <span class="md-ellipsis">
      6. Coverage filtering
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-scaffolding" class="md-nav__link">
    <span class="md-ellipsis">
      7. Scaffolding
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-annotation-with-reference" class="md-nav__link">
    <span class="md-ellipsis">
      8. Annotation with Reference
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="assembly-polishing">Assembly &amp; polishing</h1>
<p>Viralgenie offers an elaborate workflow for the assembly and polishing of viral genomes:</p>
<ol>
<li><a href="#1-de-novo-assembly">Assembly</a>: combining the results of multiple assemblers.</li>
<li><a href="#2-extension">Extension</a>: extending contigs using paired-end reads.</li>
<li><a href="#3-coverage-calculation">Coverage calculation</a>: mapping reads back to the contigs to determine coverage.</li>
<li><a href="#4-reference-matching">Reference Matching</a>: comparing contigs to a reference sequence pool.</li>
<li><a href="#5-taxonomy-guided-clustering">Taxonomy guided Clustering</a>: clustering contigs based on taxonomy and nucleotide similarity.<ul>
<li><a href="#51-pre-clustering-using-taxonomy">Pre-clustering</a>: separating contigs based on identified taxonomy-id.</li>
<li><a href="#52-actual-clustering-on-nucleotide-similarity">Actual clustering</a>: clustering contigs based on nucleotide similarity.</li>
</ul>
</li>
<li><a href="#7-scaffolding">Scaffolding</a>: scaffolding the contigs to the centroid of each bin.</li>
<li><a href="#8-annotation-with-reference">Annotation with Reference</a>: annotating regions with 0-depth coverage with the reference sequence.</li>
</ol>
<p><img alt="assembly_polishing" src="../../images/assembly_polishing.png" /></p>
<blockquote>
<p>The overall workflow of creating reference assisted assemblies can be skipped with the argument <code>--skip_assembly</code>. See the <a href="../../parameters/#assembly">parameters assembly section</a> for all relevant arguments to control the assembly steps.</p>
<p>The overall refinement of contigs can be skipped with the argument <code>--skip_polishing</code>. See the <a href="../../parameters/#polishing">parameters polishing section</a> for all relevant arguments to control the polishing steps.</p>
</blockquote>
<p>The consensus genome of all clusters are then sent to the <a href="../variant_and_refinement/">variant analysis &amp; iterative refinement</a> step.</p>
<h2 id="1-de-novo-assembly">1. De-novo Assembly</h2>
<p>Three assemblers are used, <a href="http://cab.spbu.ru/software/spades/">SPAdes</a>, <a href="https://github.com/voutcn/megahit">Megahit</a>, and <a href="https://github.com/trinityrnaseq/trinityrnaseq">Trinity</a>. The resulting contigs of all specified assemblers, are combined and processed further together.</p>
<blockquote>
<p>Modify the spades mode with <code>--spades_mode [default: rnaviral]</code> and supply specific params with <code>--spades_yml</code> or a hmm model with <code>--spades_hmm</code>.</p>
<p>Specify the assemblers to use with the <code>--assemblers</code> parameter where the assemblers are separated with a ','. The default is <code>spades,megahit,trinity</code>.</p>
</blockquote>
<p>Low complexity contigs can be filtered out using prinseq++ with the <code>--skip_contig_prinseq false</code> parameter. Complexity filtering is primarily a run-time optimisation step. Low-complexity sequences are defined as having commonly found stretches of nucleotides with limited information content (e.g. the dinucleotide repeat CACACACACA). Such sequences can produce a large number of high-scoring but biologically insignificant results in database searches. Removing these reads therefore saves computational time and resources.</p>
<h2 id="2-extension">2. Extension</h2>
<p>Contigs can be extended using <a href="https://github.com/nsoranzo/sspace_basic">SSPACE Basic</a> with the <code>--skip_sspace_basic false</code> parameter. SSPACE is a tool for scaffolding contigs using paired-end reads. It is modified from SSAKE assembler and has the feature of extending contigs using reads that are unmappable in the contig assembly step. To maximize its efficiency, consider specifying the arguments <code>--read_distance</code>, <code>--read_distance_sd</code>, and <code>--read_orientation</code>.  For more information on these arguments, see the <a href="../../parameters/#assembly">parameters assembly section</a>.</p>
<blockquote>
<p>The extension of contigs is run by default, to skip this step, use <code>--skip_sspace_basic</code>.</p>
</blockquote>
<h2 id="3-coverage-calculation">3. Coverage calculation</h2>
<p>Processed reads are mapped back against the contigs to determine the number of reads mapping towards each contig. This is done with <a href="http://bowtie-bio.sourceforge.net/bowtie2/"><code>BowTie2</code></a>,<a href="https://github.com/bwa-mem2/bwa-mem2"><code>BWAmem2</code></a> or <a href="https://github.com/lh3/bwa"><code>BWA</code></a>. This step is used to remove contig clusters that have little to no coverage downstream.</p>
<blockquote>
<p>Specify the mapper to use with the <code>--mapper</code> parameter. The default is <a href="https://github.com/bwa-mem2/bwa-mem2"><code>BWAmem2</code></a>. To skip contig filtering specify <code>--perc_reads_contig 0</code>.</p>
</blockquote>
<h2 id="4-reference-matching">4. Reference Matching</h2>
<p>The newly assembled contigs are compared to a reference sequence pool (<code>--reference_pool</code>) using a <a href="https://www.ncbi.nlm.nih.gov/books/NBK153387/">BLASTn search</a>. This process not only helps annotate the contigs but also assists in linking together sets of contigs that are distant within a single genome. Essentially, it aids in identifying contigs belonging to the same genomic segment and choosing the right reference for scaffolding purposes.</p>
<p>The top 5 hits for each contig are combined with the de novo contigs and sent to the clustering step.</p>
<blockquote>
<p>The reference pool can be specified with the <code>--reference_pool</code> parameter. The default is the latest clustered <a href="https://rvdb.dbi.udel.edu/">Reference Viral DataBase (RVDB)</a>.</p>
</blockquote>
<h2 id="5-taxonomy-guided-clustering">5. Taxonomy guided Clustering</h2>
<p>The clustering workflow of contigs consists of 2 steps, the <a href="#51-pre-clustering-using-taxonomy">pre-clustering using taxonomy</a> and
<a href="#52-actual-clustering-on-nucleotide-similarity">actual clustering on nucleotide similarity</a>. The taxonomy guided clustering is used to separate contigs based on taxonomy and nucleotide similarity.</p>
<pre class="mermaid"><code>graph LR;
    A[Contigs] --&gt; B["`**Pre-clustering**`"];
    B --&gt; C["`**Actual clustering**`"];</code></pre>
<h3 id="51-pre-clustering-using-taxonomy">5.1 Pre-clustering using taxonomy</h3>
<p>The contigs along with their references have their taxonomy assigned using <a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> and <a href="https://kaiju.binf.ku.dk/">Kaiju</a>.</p>
<blockquote>
<p>The default databases are the same ones used for read classification:
- Kraken2: viral refseq database, <code>--kraken2_db</code>
- Kaiju: clustered <a href="https://rvdb.dbi.udel.edu/">RVDB</a>, <code>--kaiju_db</code></p>
</blockquote>
<p>As Kaiju and Kraken2 can have different taxonomic assignments, an additional step is performed to resolve potential inconsistencies in taxonomy and to identify the taxonomy of the contigs. This is done with a custom script that is based on <code>KrakenTools extract_kraken_reads.py</code> and <code>kaiju-Merge-Outputs</code>.</p>
<pre class="mermaid"><code>graph LR;
    A[Contigs] --&gt; B["`**Kraken2**`"];
    A --&gt; C["`**Kaiju**`"];
    B --&gt; D[Taxon merge resolving];
    C --&gt; D;
    D --&gt; E["Taxon filtering"];
    E --&gt; F["Taxon simplification"];</code></pre>
<div class="admonition tip annotate">
<p class="admonition-title">Having complex metagenomic samples?</p>
<p>The pre-clustering step can be used to simplify the taxonomy of the contigs, let <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi">NCBI's taxonomy browser</a> help you identify taxon-id's for simplification. The simplification can be done in several ways:</p>
<ul>
<li>Make sure your contamination database is up to date and removes the relevant taxa.</li>
<li>Exclude unclassified contigs with <code>--keep_unclassified false</code> parameter.</li>
<li>Simplify the taxonomy of the contigs to a higher rank using <code>--precluster_simplify_taxa</code> parameter (1).</li>
<li>Specify the taxa to include or exclude with <code>--precluster_include_children</code>(2), <code>--precluster_include_parents</code>(3), <code>--precluster_exclude_children</code>, <code>--precluster_exclude_parents</code>, <code>--precluster_exclude_taxa</code> parameters.</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Providing lists to nextflow is done by encapsulating values with <code>"</code> and separating them with a space. For example: <code>--precluster_exclude_taxa "taxon1 taxon2 taxon3"</code>.</p>
</div>
</div>
<ol>
<li>
<p>Options here are 'species', 'genus', 'family', 'order', 'class', 'phylum', 'kingdom' or 'superkingdom'.</p>
</li>
<li>
<p><code>--precluster_include_children</code> <strong>"genus1"</strong> :</p>
<p><pre class="mermaid"><code>graph TD;
    A[family] -.- B["genus1 (included)"];
    A -.- C[genus2];
    B --- D[species1];
    B --- E[species2];
    C -.- F[species3];</code></pre>
Dotted lines represent exclusion of taxa.</p>
</li>
<li>
<p><code>--precluster_include_parents</code> <strong>"species3"</strong> :</p>
<p><pre class="mermaid"><code>graph TD;
    A["family (included)"] -.- B["genus1"]
    A --- C[genus2]
    B -.- D[species1]
    B -.- E[species2]
    C --- F[species3]</code></pre>
Dotted lines represent exclusion of taxa.</p>
</li>
</ol>
<blockquote>
<p>The pre-clustering step will be run by default but can be skipped with the argument <code>--skip_preclustering</code>. Specify which classifier to use with <code>--precluster_classifiers</code> parameter. The default is <code>kaiju,kraken2</code>. Contig taxon filtering is still enabled despite not having to solve for inconsistencies if only Kaiju or Kraken2 is run.</p>
</blockquote>
<h3 id="52-actual-clustering-on-nucleotide-similarity">5.2 Actual clustering on nucleotide similarity</h3>
<p>The clustering is performed with one of the following tools:</p>
<ul>
<li><a href="https://sites.google.com/view/cd-hit"><code>cdhitest</code></a></li>
<li><a href="https://github.com/torognes/vsearch/wiki/Clustering"><code>vsearch</code></a></li>
<li><a href="https://github.com/soedinglab/MMseqs2/wiki#linear-time-clustering-using-mmseqs-linclust"><code>mmseqs-linclust</code></a></li>
<li><a href="https://github.com/soedinglab/MMseqs2/wiki#cascaded-clustering"><code>mmseqs-cluster</code></a></li>
<li><a href="https://github.com/AnantharamanLab/vRhyme"><code>vRhyme</code></a></li>
<li><a href="https://github.com/marbl/Mash"><code>mash</code></a></li>
</ul>
<p>These methods all come with their own advantages and disadvantages. For example, cdhitest is very fast but cannot be used for large viruses &gt;10Mb and similarity threshold cannot go below 80% which is not preferable for highly diverse RNA viruses. Vsearch is slower but accurate. Mmseqs-linclust is the fastest but tends to create a large amount of bins. Mmseqs-cluster is slower but can handle larger datasets and is more accurate. vRhyme is a new method that is still under development but has shown promising results but can sometimes not output any bins when segments are small. Mash is a very fast comparison method is linked with a custom script that identifies communities within a network.</p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>When pre-clustering is performed, it is recommended to set a lower identity_threshold (60-70% ANI) as the new goal becomes to separate genome segments within the same bin.</p>
</div>
<blockquote>
<p>The clustering method can be specified with the <code>--clustering_method</code> parameter. The default is <code>cdhitest</code>.</p>
<p>The network clustering method for <code>mash</code> can be specified with the <code>--network_clustering</code> parameter. The default is <code>connected_components</code>, alternative is <a href="https://www.nature.com/articles/s41598-019-41695-z"><code>leiden</code></a>.</p>
<p>The similarity threshold can be specified with the <code>--similarity_threshold</code> parameter. The default is <code>0.85</code>.</p>
</blockquote>
<h2 id="6-coverage-filtering">6. Coverage filtering</h2>
<p>The coverage of the contigs is calculated using the same method as in the <a href="#3-coverage-calculation">coverage calculation step</a>. A cumulative sum is taken across the contigs from every assembler. If these cumulative sums are above the specified <code>--perc_reads_contig</code> parameter, the contig is kept. If all cumulative sums are below the specified parameter, the contig is removed.</p>
<div class="admonition info annotate">
<p class="admonition-title">Show me an example how it works</p>
<p>If the <code>--perc_reads_contig</code> is set to <code>5</code>, the cumulative sum of the contigs from every assembler is calculated. For example:</p>
<ul>
<li>Cluster 1: the cumulative sum of the contigs from SPAdes is 0.6, Megahit is 0.5, the cluster is kept.</li>
<li>Cluster 2: the cumulative sum of the contigs from SPAdes is 0.1, Megahit is 0.1, the cluster is removed.</li>
<li>Cluster 3: the cumulative sum of the contigs from SPAdes is 0.5, Megahit is 0, the cluster is kept.</li>
</ul>
</div>
<blockquote>
<p>The default is <code>5</code> and can be specified with the <code>--perc_reads_contig</code> parameter.</p>
</blockquote>
<h2 id="7-scaffolding">7. Scaffolding</h2>
<p>After classifying all contigs and their top BLAST hits into distinct clusters or bins, the contigs are then scaffolded to the centroid of each bin. Any external references that are not centroids of the cluster are subsequently removed to prevent further bias. All members of the cluster are consequently mapped towards their centroid with <a href="https://github.com/lh3/minimap2">Minimap2</a> and consensus is called using <a href="https://andersen-lab.github.io/ivar/html/manualpage.html">iVar-consensus</a>.</p>
<h2 id="8-annotation-with-reference">8. Annotation with Reference</h2>
<p>Regions with 0-depth coverage are annotated with the reference sequence. This is done with a <a href="https://github.com/Joon-Klaps/viralgenie/blob/dev/bin/lowcov_to_reference.py">custom script</a> that uses the coverage of the de novo contigs towards the reference sequence to identify regions with 0-depth coverage. The reference sequence is then annotated to these regions.</p>
<blockquote>
<p>This step can be skipped using <code>--skip_hybrid_consensus</code> parameter.</p>
</blockquote>









  



  <form class="md-feedback" name="feedback" hidden>
    <fieldset>
      <legend class="md-feedback__title">
        Was this page helpful?
      </legend>
      <div class="md-feedback__inner">
        <div class="md-feedback__list">
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page was helpful" data-md-value="1">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12 21.35-1.45-1.32C5.4 15.36 2 12.27 2 8.5 2 5.41 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.08C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.41 22 8.5c0 3.77-3.4 6.86-8.55 11.53z"/></svg>
            </button>
          
            <button class="md-feedback__icon md-icon" type="submit" title="This page could be improved" data-md-value="0">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12 21.35-1.45-1.32C5.4 15.36 2 12.27 2 8.5 2 5.41 4.42 3 7.5 3c.67 0 1.32.12 1.94.33L13 9.35l-4 5zM16.5 3C19.58 3 22 5.41 22 8.5c0 3.77-3.4 6.86-8.55 11.53L12 21.35l-1-7 4.5-5-2.65-5.08C13.87 3.47 15.17 3 16.5 3"/></svg>
            </button>
          
        </div>
        <div class="md-feedback__note">
          
            <div data-md-value="1" hidden>
              
              
                
              
              Thanks for your feedback, I like you too!
            </div>
          
            <div data-md-value="0" hidden>
              
              
                
              
              Thanks for your feedback! Help us improve this page by submitting an <a href="https://github.com/Joon-Klaps/viralgenie/issues/new/?labels=website%2Cdocumentation&template=website_feedback.yml&title=[Website feedback]+Assembly%20%26%20Polishing" target="_blank" rel="noopener">issue</a>.
            </div>
          
        </div>
      </div>
    </fieldset>
  </form>


                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../metagenomic_diversity/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Metagenomic diversity">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                Metagenomic diversity
              </div>
            </div>
          </a>
        
        
          
          <a href="../variant_and_refinement/" class="md-footer__link md-footer__link--next" aria-label="Next: Variant analysis &amp; Iterative refinement">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Variant analysis & Iterative refinement
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://twitter.com/LemeyLab" target="_blank" rel="noopener" title="twitter.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8l164.9-188.5L26.8 48h145.6l100.5 132.9zm-24.8 373.8h39.1L151.1 88h-42z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/Joon-Klaps" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
      <div class="md-consent" data-md-component="consent" id="__consent" hidden>
        <div class="md-consent__overlay"></div>
        <aside class="md-consent__inner">
          <form class="md-consent__form md-grid md-typeset" name="consent">
            

  
    
  


  
    
  



  


<h4>Cookie Consent</h4>
<p>We use cookies to recognize your repeated visits and preferences, as well as to measure the effectiveness of our documentation and whether users find what they're searching for. With your consent, you're helping us to make our documentation better.</p>
<input class="md-toggle" type="checkbox" id="__settings" >
<div class="md-consent__settings">
  <ul class="task-list">
    
      
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="analytics" checked>
          <span class="task-list-indicator"></span>
          Google Analytics
        </label>
      </li>
    
      
      
        
        
      
      <li class="task-list-item">
        <label class="task-list-control">
          <input type="checkbox" name="github" checked>
          <span class="task-list-indicator"></span>
          GitHub
        </label>
      </li>
    
  </ul>
</div>
<div class="md-consent__controls">
  
    
      <button class="md-button md-button--primary">Accept</button>
    
    
    
  
    
    
    
      <label class="md-button" for="__settings">Manage settings</label>
    
  
</div>
          </form>
        </aside>
      </div>
      <script>var consent=__md_get("__consent");if(consent)for(var input of document.forms.consent.elements)input.name&&(input.checked=consent[input.name]||!1);else"file:"!==location.protocol&&setTimeout((function(){document.querySelector("[data-md-component=consent]").hidden=!1}),250);var form=document.forms.consent;for(var action of["submit","reset"])form.addEventListener(action,(function(e){if(e.preventDefault(),"reset"===e.type)for(var n of document.forms.consent.elements)n.name&&(n.checked=!1);__md_set("__consent",Object.fromEntries(Array.from(new FormData(form).keys()).map((function(e){return[e,!0]})))),location.hash="",location.reload()}))</script>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["announce.dismiss", "content.code.annotate", "content.code.copy", "header.autohide", "navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.indexes", "navigation.tabs.sticky", "navigation.top", "navigation.tracking", "search.share", "search.suggest", "toc.follow"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
        <script src="https://unpkg.com/tablesort@5.3.0/dist/tablesort.min.js"></script>
      
        <script src="../../javascripts/tablesort.js"></script>
      
        <script src="../../javascripts/katex.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/katex.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.7/contrib/auto-render.min.js"></script>
      
    
  </body>
</html>