Changes in module 'nf-core/quast'
--- modules/nf-core/quast/main.nf
+++ modules/nf-core/quast/main.nf
@@ -1,6 +1,7 @@
 process QUAST {
     tag "$meta.id"
     label 'process_medium'
+    errorStrategy { task.exitStatus == 4 ? 'ignore' : sleep(10 * 200 as long); return 'retry' } // can fail if mpileup empty
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?

************************************************************
