# Output

!!! Warning "INCOMPLETE, YET TO BE WRITTEN"


## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

!!! Tip
    A global, partly, random prefix can be created using the argument `--prefix <string>`. The following string will then be used as a prefix to all output files.
    ```java
    "<prefix_string>_<date>_<pipeline_version>_<workflow_runName>"
    ```

## Preprocessing

All output files of the preprocessing steps can be found in the directory `preprocessing/`.
### FastQC

???- abstract "Output files"
    -   `fastqc/{raw,trim,host}`
        -   `*_fastqc.html`: FastQC report containing quality metrics.
        -   `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

!!! Tip
    The FastQC plots displayed in the MultiQC report shows _untrimmed_, _trimmed_ and _host filtered_ reads. Make sure to check the section titles for the correct set of reads.

### fastp

[fastp](https://github.com/OpenGene/fastp) is a FASTQ pre-processing tool for quality control, trimmming of adapters, quality filtering and other features.

???- abstract "Output files"

    - `fastp/`
        - `report/<sample_id>.*{html,json}`: report files in different formats.
        - `log/<sample_id>.*{html,json}`: log files.
        - `<sample_id>.fastp.fastq.gz`: file with the trimmed fastq reads.
        - `fail/<sample_id>.fail.fastq.gz`: file with reads that didn't suffice quality controls.

By default viralgenie will only provide the report and log files if fastp is selected. The trimmed reads can be saved by specifying `--save_intermediate_reads` or `--save_final_reads 'trimming'`. Similarly, the saving of the output reads can be enabled with `--save_trimmed_fail`.

### Trimmomatic
[Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) is a FASTQ pre-processing tool for quality control, trimmming of adapters, quality filtering and other features.

???- abstract "Output files"

    - `trimmomatic/`
        - `*.fastq.gz`: file with the trimmed fastq reads.
        - `log/<sample_id>.*{html,txt,zip}`: log files generated by trimmomatic.

By default viralgenie will only provide the report and log files if Trimmomatic is selected. The trimmed reads can be saved by specifying `--save_intermediate_reads` or `--save_final_reads 'trimming'`.

### BBDuk

[BBDuk](https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/bbduk-guide/) stands for Decontamination Using Kmers. BBDuk was developed to combine most common data-quality-related trimming, filtering, and masking operations into a single high-performance tool.

It is used in viralgenie for complexity filtering using different algorithms. This means that it will remove reads with low sequence diversity (e.g. mono- or dinucleotide repeats).

???- abstract "Output files"

    - `bbduk/`
        - `log/<sample_id>.bbduk.log`: log file containing filtering statistics
        - `<sample_id>.fastq.gz`: resulting FASTQ file without low-complexity reads

By default viralgenie will only provide the log files of bbduk. The filtered reads can be saved by specifying `--save_intermediate_reads` or `--save_final_reads 'complexity'`.


### Hostremoval-Kraken2

[Kraken2](https://ccb.jhu.edu/software/kraken2/) is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps -mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.

???- abstract "Output files"

    - `hostremoval-kraken2/`
        * `<sample_id>_kraken2_host.report.txt`: A profile of the aligned reads to a given host contamination database.
        * `<sample_id>_kraken2_host.unclassified*.fastq.gz`: resulting FASTQ file with reads that don't have any matches to the given host contamination database.


By default viralgenie will only provide the log files of kraken2 which are visualised in [Multiqc](#multiqc). The filtered reads can be saved by specifying `--save_intermediate_reads` or `--save_final_reads 'host'`.

## Metagenomic Diversity

The results of the metagenomic diversity analysis are stored in the directory `metagenomic_diversity/`. Results are also visualised in the MultiQC report.
![Multiqc kaiju example](images/mqc_kaiju.png)

### Kraken2
[Kraken](https://ccb.jhu.edu/software/kraken2/) is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps -mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.

???- abstract "Output files"

    - `metagenomic_diversity/kraken2/`

        - `<sample_id>.report.txt`:A Kraken2 report that summarises the fraction abundance, taxonomic ID, number of Kmers, taxonomic path of all the hits in the Kraken2 run for a given sample. Will be 6 column rather than 8 if `--save_minimizers` specified.
        - `<sample_id>_kraken2_host.unclassified*.fastq.gz`: resulting FASTQ file with reads that don't have any matches to the given host contamination database.
        - `<sample_id>.classified.fastq.gz`: FASTQ file containing all reads that had a hit against a reference in the database for a given sample.
        - `<sample_id>.unclassified.fastq.gz`: FASTQ file containing all reads that did not have a hit in the database for a given sample.
        - `<sample_id>.classifiedreads.txt`: A list of read IDs and the hits each read had against the database for a given sample.

> By default viralgenie will provide any classified or unclassified fastq files, specify this with --kraken2_save_reads. Similarly for the classifiedreads table, specify this with --kraken2_save_readclassification.

### Kaiju
[Kaiju](https://kaiju.binf.ku.dk/) is a program for sensitive taxonomic classification of high-throughput sequencing reads from metagenomic data. It is based on the Burrows-Wheeler transform and the lowest common ancestor algorithm.

???- abstract "Output files"

    - `metagenomic_diversity/kaiju/`
        - `<sample_id>.tsv`: Raw output from Kaiju with taxonomic rank, read ID and taxonic ID
        - `<sample_id>.txt`: A summary of the taxonomic classification of the reads in the sample.

### Krona
[Krona](https://github.com/marbl/Krona/wiki) is a hierarchical data visualisation tool that can be used to visualise the taxonomic classification of metagenomic data.

???- abstract "Output files"

    - `metagenomic_diversity/krona/`
        - `<kaiju|kraken2>_.html`: A HTML file containing the Krona visualisation of the taxonomic classification of the reads in the sample.


![Krona example COVID](images/krona_covid.png){: .center}

## Assembly & Polishing

The results of the assembly processes & polishing are stored in the directory `assembly/`.

Multiple intermediate files can be genarated during the assembly process, some of them might not always be interesting to have. For this reason, there is an option to save the intermediate files with the `--save_intermediate_polishing` argument which is by default off.

### Assemblers

Multiple assemblers [spades, trinity, megahit] can be used which have their results combined. Each assembler has its own directory in the `assembly/assemblers` directory, where there will be a subfolder for the contigs and the QC results from quast.

???- abstract "Output files"

    -   `assemblers/`
        -   `spades/<spades_mode>/`
            -   `contigs/<sample_id>_spades.fa.gz`: Contigs generated by SPAdes.
            -  `log/<sample_id>_spades.log`: Directory containing the log file of the spades run.
            -   `quast/<sample_id>_spades.tsv`: Directory containing the QUAST report.
        -   `trinity/`
            -   `contigs/<sample_id>_trinity.fa.gz`: Contigs generated by Trinity.
            -   `quast/<sample_id>_trinity.tsv`: Directory containing the QUAST report.
        -   `megahit/`
            -   `contigs/<sample_id>_megahit.fa.gz`: Contigs generated by Megahit.
            -   `quast/<sample_id>_megahit.tsv`: Directory containing the QUAST report.

Quast results are also summarised and plotted in the MultiQC report.

![MultiQC - QUAST assembly statistics](images/mqc_quast.png){: .center}

Finally, the results of the assemblers are combined and stored in the `tools_combined/` directory.

???- abstract "Output files"

    - `assemblers`
        - `tools_combined/<sample_id>.combined.fa` : Contigs generated by combining the results of the assemblers.

### BLAST

[BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) is a sequence comparison tool that can be used to compare a query sequence against a database of sequences. In viralgenie, BLAST is used to compare the contigs generated by the assemblers to a database of viral sequences.

By default, viralgenie will only provide the BLAST results in a tabular format. It will have selected only for the top five hits and will also have filtered version where it will only include hits with an e-value of 0.01 or lower, a bitscore of 50 or higher and a alignment percentage of 0.80 or higher.

???- info "Column names"

    - qseqid
    - sseqid
    - stitle
    - pident
    - qlen
    - slen
    - length
    - mismatch
    - gapopen
    - qstart
    - qend
    - sstart
    - send
    - evalue
    - bitscore


???- abstract "Output files"

    - `polishing/`
        - `blast/<sample_id>_filter.tsv`: Filtered BLAST results in tabular format.
        - `intermediate/blast/filtered-sequences/<sample_id>_withref.fa`: Contigs with the blast hit sequence in a fasta file.
        - `intermediate/blast/hits/<sample_id>.txt`: unfiltered BLAST results in tabular format.

> By default viralgenie will only provide the filtered blast.txt file. The intermediate files can be saved by specifying `--save_intermediate_polishing`.

### Preclustering - Kaiju & Kraken2

[Kaiju](https://kaiju.binf.ku.dk/) is a program for sensitive taxonomic classification of high-throughput sequencing reads from metagenomic data. It is based on the Burrows-Wheeler transform and the lowest common ancestor algorithm.

[Kraken2](https://ccb.jhu.edu/software/kraken2/) is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences. Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database. That database maps -mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.

???- abstract "Output files"

    - `polishing/intermediate/precluster`
        - `kaiju/<sample_id>_kaiju.tsv`: Raw output from Kaiju with taxonomic rank, read ID and taxonic ID
        - `kraken2/<sample_id>_kraken2_reports.txt`: A Kraken2 report that summarises the fraction abundance, taxonomic ID, number of Kmers, taxonomic path of all the hits in the Kraken2 run for a given sample. Will be 6 column rather than 8 if `--save_minimizers` specified.
        - `kraken2/<sample_id>_kraken2.classifiedreads.txt`: A list of read IDs and the hits each read had against the database for a given sample.
        - `merged_classifications/<sample_id>_merged.tsv`: Taxonomy merged based on the specified strategy with the columns being taxonomic rank, read ID and taxonic ID.
        - `sequences/<sample_id>/<sample_id>_taxid<taxonic ID>.fa`: Fasta file with the contigs that were classified to that specific taxonomic ID.

> By default viralgenie will not provide any preclustering file. The intermediate files can be saved by specifying `--save_intermediate_polishing`.


### Clustering

The output files of each clustering method is directly put in te `assembly/polishing` directory. With the exception of a summary file that is generated by the pipeline for each cluster with the size of the cluster, the centroid, ... .

???- abstract "Output files"

    - `polishing/intermediate/cluster/`
        - `<sample_id>/<sample_id>_cl#.json`: A json file containign information that is used by viralgenie internally like clusterID, centroid, members, cluster size, if the reference is external or a denovo contig.
        - `<sample_id>/<sample_id>.summary_mqc.tsv`: A tabular file with comments used for [Multiqc](#multiqc) with statistics on the number of identified clusters in a sample
        - `<sample_id>/<sample_id>.clusters.tsv`: A tabular file with metadata on all clusters in a samples. It's the json file of all clusters in a table format.

!!! Tip

    Whenever there is a 'cl#' in the file name, it refers to the cluster number of that sample.

> By default viralgenie will not provide any clustering overview files. The intermediate files can be saved by specifying `--save_intermediate_polishing`.

#### CD-HIT-EST
[CD-HIT](https://github.com/weizhongli/cdhit) is a very fast, widely used program for clustering and comparing protein or nucleotide sequences.

???- abstract "Output files"

    - `polishing/cdhit/`
        - `<sample_id>/<sample_id>.fa.clstr`: A cluster file containing the clustering information. where  ">" starts a new cluster, a "*" at the end means that this sequence is the representative or centroid of this cluster, and a "%" is the identity between this sequence and the representative
        - `<sample_id>/<sample_id>.fa`: A fasta file containing the centroid sequence.

#### vsearch-cluster
[vsearch](https://github.com/torognes/vsearch) implements a single-pass, greedy centroid-based clustering algorithm, similar to the algo- rithms implemented in usearch, DNAclust and sumaclust for example. The output has to be in the `--uc` format or else the pipeline will not be able to process the output.

???- abstract "Output files"

    - `polishing/vsearch/`
        - `<sample_id>/<sample_id>.tsv.gz`: A cluster file containing the clustering information.

???- info "vsearch -uc columns"

    1. Entry (S, H, or C):
    2. Record type: S, H, or C.
    3. Cluster number (zero-based).
    4. Centroid length (S), query length (H), or cluster size (C).
    5. Percentage of similarity with the centroid sequence (H), or set to ’*’ (S, C).
    6. Match orientation + or - (H), or set to ’*’ (S, C). Not used, always set to ’*’ (S, C) or to zero (H).
    7. Not used, always set to ’*’ (S, C) or to zero (H).
    8. Set to ’*’ (S, C) or, for H, compact representation of the pairwise alignment using the CIGAR format (Compact Idiosyncratic Gapped Alignment Report): M (match/mismatch), D (deletion), and I (insertion). The equal sign ’=’ indicates that the query is identical to the centroid sequence.
    9. Label of the query sequence (H), or of the centroid sequence (S, C). 10. Label of the centroid sequence (H), or set to ’*’ (S, C).

#### MMseqs2

[MMseqs2](https://github.com/soedinglab/MMseqs2/wiki) is a software suite to search and cluster huge protein and nucleotide sequence sets. The cascaded clustering workflow (`mmseqs-cluster`) first runs linclust, the linear-time clustering module of mmseqs (`mmseqs-linclust`), that can produce clustering’s down to 50% sequence identity in very short time.

???- abstract "Output files"

    - `polishing/`
        - `mmseqs2/<sample_id>/<sample_id>.tsv`: A cluster file containing the clustering information. Where the first column is the cluster representative and the second column the member.
        - intermediate/mmseqs/clustered_db/<sample_id>*`: A MMseqs2 database of the clustered sequences.
        - intermediate/mmseqs/sequence_db/<sample_id>*`: A MMseqs2 database of the input sequences (contigs + blast hits).

#### vRhyme

[vRhyme](https://github.com/AnantharamanLab/vRhyme) is a multi-functional tool for binning virus genomes from metagenomes. vRhyme functions by utilizing coverage variance comparisons and supervised machine learning classification of sequence features to construct viral metagenome-assembled genomes (vMAGs).

???- abstract "Output files"

    - `polishing/vrhyme`
        - `<sample_id>/vRhyme_best_bins.#.membership.tsv`: scaffold membership of best bins
        - `<sample_id>/vRhyme_best_bins.#.summary.tsv`: summary stats of best bins

#### Mash
[Mash](https://github.com/marbl/Mash) calculates the distance between two sequences based on the jaccard distance. The Mash distance can be quickly computed from the size-reduced sketches alone, yet produces a result that strongly correlates with alignment-based measures such as the Average Nucleotide Identity (ANI).

???- abstract "Output files"

    - `polishing/mash`
        - `<sample_id>/dist/*.tsv`: A distance matrix of the genomes with ANI
        - `<sample_id>/cluster/*.tsv`: A table where the first column represents the contig/genome and the second column it's corresponding cluster.
        - `<sample_id>/visual/*.png `: A visualisation of the network.

The network of a triple segmented Hazara virus looks like this, each node represents a contig colored on cluster. The edge represents that the ANI is higher then the specified `--identity_threshold`.


![mash HAZV example image](images/mash_HAZV_example.png){: .center}

!!! info "What are those names?"
    Most assemblers tend to give each contig name a specific prefix. For example,

    - Trinity: `'TRINITY_...'`
    - SPAdes: `'NODE_...'`
    - Megahit: `'k\d{3}_...'`

    Based on these prefixes viralgenie separates external references from denovo contigs. If any assemblers are added, consider specifying a specific regex for `--assembler_patterns`.

### Minimap2

[Minimap2](https://github.com/lh3/minimap2) is a versatile sequence alignment program that aligns larger DNA or mRNA sequences against a large reference database.

???- abstract "Output files"

    - `polishing/scaffolding/<sample_id>/minimap`
        - `<sample_id>_cl#.bam`: A BAM file containing the alignment of contigs to the centroid.
        - `<sample_id>_cl#.mmi`: The centroid index file.

> By default viralgenie will not provide the minimap output files. The intermediate files can be saved by specifying `--save_intermediate_polishing`.

### iVar contig consensus
[iVar](https://andersen-lab.github.io/ivar/html/manualpage.html#autotoc_md19) is a computational method for calling consensus sequences from viral populations.

???- abstract "Output files"

    - `polishing/scaffolding/<sample_id>/ivar`
        - `<sample_id>_cl#_consensus.fa`: A fasta file containing the consensus sequence of the cluster.
        - `<sample_id>_cl#_consensus.mpileup`: A mpileup file containing depth at each position of the consensus sequence.
        - `<sample_id>_cl#_consensus.qual.txt`: A text file contianing the quality of each base in the consensus sequence.
        - `hybrid-<sample_id>_cl#_consensus.fa`: A fasta file containing the hybrid consensus sequence of the cluster and the reference.

> By default viralgenie will not provide the iVar output files. The intermediate files can be saved by specifying `--save_intermediate_polishing`.

!!! Info
    The hybrid consensus is generated by mapping the contigs to the reference and then calling the consensus sequence. This is done to fill in the gaps in the contigs with the reference sequence, if there are no postions with 0 coverage there will not be a hybrid consensus and the output from iVar will be used.

## Variant Calling & Iterative Refinement




## MultiQC

???- abstract "Output files"
    -   `multiqc/`
        -   `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
        -   `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
        -   `multiqc_plots/`: directory containing static images from the report in various formats.

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

## Pipeline information

???- abstract "Output files"

    -   `pipeline_info/`
        -   Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
        -   Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
        -   Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
        -   Parameters used by the pipeline run: `params.json`.

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
