Changes in module 'nf-core/prinseqplusplus'
--- modules/nf-core/prinseqplusplus/main.nf
+++ modules/nf-core/prinseqplusplus/main.nf
@@ -8,14 +8,14 @@
         'biocontainers/prinseq-plus-plus:1.2.3--hc90279e_1' }"
 
     input:
-    tuple val(meta), path(reads)
+    tuple val(meta), path(reads), path(fasta)
 
     output:
-    tuple val(meta), path("*_good_out*.fastq.gz")                  , emit: good_reads
-    tuple val(meta), path("*_single_out*.fastq.gz"), optional: true, emit: single_reads
-    tuple val(meta), path("*_bad_out*.fastq.gz")   , optional: true, emit: bad_reads
-    tuple val(meta), path("*.log")                                 , emit: log
-    path "versions.yml"                                            , emit: versions
+    tuple val(meta), path("*_good_out*")    , emit: good_reads
+    tuple val(meta), path("*_single_out*")  , optional: true, emit: single_reads
+    tuple val(meta), path("*_bad_out*")     , optional: true, emit: bad_reads
+    tuple val(meta), path("*.log")          , emit: log
+    path "versions.yml"                     , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
@@ -23,39 +23,22 @@
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def fastqs = reads ? meta.single_end ? "-fastq ${reads}" : "-fastq ${reads[0]} -fastq2 ${reads[1]}" : ''
+    def fasta = fasta ? "-fasta ${fasta}" : ''
 
-    if (meta.single_end) {
-        """
-        prinseq++ \\
-            -threads $task.cpus \\
-            -fastq ${reads} \\
-            -out_name ${prefix} \\
-            -out_gz \\
-            -VERBOSE 1 \\
-            $args \\
-            | tee ${prefix}.log
+    """
+    prinseq++ \\
+        -threads $task.cpus \\
+        ${fastqs} \\
+        ${fasta} \\
+        -out_name ${prefix} \\
+        -VERBOSE 1 \\
+        $args \\
+        | tee ${prefix}.log
 
-        cat <<-END_VERSIONS > versions.yml
-        "${task.process}":
-            prinseqplusplus: \$(echo \$(prinseq++ --version | cut -f 2 -d ' ' ))
-        END_VERSIONS
-        """
-    } else {
-        """
-        prinseq++ \\
-            -threads $task.cpus \\
-            -fastq ${reads[0]} \\
-            -fastq2 ${reads[1]} \\
-            -out_name ${prefix} \\
-            -out_gz \\
-            -VERBOSE 1 \\
-            $args \\
-            | tee ${prefix}.log
-
-        cat <<-END_VERSIONS > versions.yml
-        "${task.process}":
-            prinseqplusplus: \$(echo \$(prinseq++ --version | cut -f 2 -d ' ' ))
-        END_VERSIONS
-        """
-    }
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        prinseqplusplus: \$(echo \$(prinseq++ --version | cut -f 2 -d ' ' ))
+    END_VERSIONS
+    """
 }

************************************************************
